{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/delphy_workflows/blob/main/delphy_workflow.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "# Delphy workflow\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 1. Select your virus of interest and apply filters to the genomes downloaded from NCBI virus"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "virus = '11070'                   # Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid)\n",
        "accession = False                 # If 'virus' is an NCBI accession instead of a taxon (e.g. 'NC_045512.2'), set this to True\n",
        "\n",
        "# Commonly used filtering options (set any filter to None to turn off the filter):\n",
        "host = 'homo sapiens'             # Example: 'homo sapiens' (alternatively: use the host_taxid filter below)\n",
        "min_seq_length = None             # Example: 6252\n",
        "max_seq_length = None             # Example: 7815\n",
        "\n",
        "has_proteins = None               # Example: 'GPC' or 'L' or ['GPC', 'L'] (also accepts genes or segments)\n",
        "proteins_complete = None          # Set to True if the proteins/genes/segments in has_proteins should be marked 'complete'\n",
        "\n",
        "geographic_location = None        # Example: 'South_Africa' or 'Germany'\n",
        "min_collection_date = None        # Example: '2000-01-01'\n",
        "max_collection_date = None        # Example: '2014-12-04'\n",
        "max_ambiguous_chars = None        # Example: 10\n",
        "\n",
        "# Additional filtering options:\n",
        "min_gene_count = None             # Example: 1\n",
        "max_gene_count = None             # Example: 40\n",
        "nuc_completeness = 'complete'     # 'partial' or 'complete'\n",
        "host_taxid = None                 # Example: 9443 (NCBI Taxonomy ID of all primates)\n",
        "lab_passaged = None               # True or False (indicates whether the virus sequence has been passaged in a laboratory setting)\n",
        "geographic_region = None          # Example: 'Africa' or 'Europe'\n",
        "submitter_country = None          # Example: 'South_Africa' or 'Germany'\n",
        "annotated = None                  # True or False (indicates whether the virus genome sequence should be annotated)\n",
        "source_database = None            # Example: 'GenBank' or 'RefSeq'\n",
        "min_release_date = None           # Example: '2000-01-01'\n",
        "max_release_date = None           # Example: '2014-12-04'\n",
        "min_mature_peptide_count = None   # Example: 2\n",
        "max_mature_peptide_count = None   # Example: 15\n",
        "min_protein_count = None          # Example: 2\n",
        "max_protein_count = None          # Example: 15"
      ],
      "metadata": {
        "id": "HLEVMl9WBgzh"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  1) Click on the folder icon on the left  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload)  \n",
        "  3) Specify the name of your file(s) here:"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "fasta_file = None        # Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'\n",
        "\n",
        "# If the metadata is the same for all sequences in your fasta file, enter the metadata here\n",
        "# You have to enter a Collection Date. In addition, you can add as many additional columns as you wish, e.g. \"Geo Location\": \"South Korea\".\n",
        "# Use NCBI column names where applicable, e.g. see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus\n",
        "metadata = {\n",
        "    \"Collection Date\": \"YYYY-MM-DD\",\n",
        "    \"Extra column 1\": \"Value\",\n",
        "    # ...\n",
        "}\n",
        "\n",
        "# Alternative: Upload a csv file containing the metadata\n",
        "# This file has to include at least a \"Accession\" and a \"Collection Date\" column\n",
        "# Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the fasta\n",
        "# Use NCBI column names where applicable, e.g. see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus\n",
        "metadata_csv = None       # Example: 'my_metadata.csv'"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 3. Click on 'Runtime' -> 'Run all' and lean back\n",
        "___"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GAljcjWNxssy"
      },
      "source": [
        "### Installing gget:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "DagUmlWZvTfz",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "17de9046-600a-4717-f927-fc2115d8e990"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m25.2/25.2 MB\u001b[0m \u001b[31m15.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m18.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for gget (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        }
      ],
      "source": [
        "# After the release, this will just be: pip install gget\n",
        "!pip install -q mysql-connector-python==8.0.29 biopython\n",
        "!pip install -q --log log git+https://github.com/pachterlab/gget.git@delphy_dev\n",
        "\n",
        "import gget"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GLnuvH8Xx-jr"
      },
      "source": [
        "Full descriptions for the filtering options:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "ZihmEChGx9j6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "dbf119c3-8eb2-48ab-f30e-98416147bb1a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on function ncbi_virus in module gget.gget_ncbi_virus:\n",
            "\n",
            "ncbi_virus(virus, accession=False, outfolder=None, host=None, min_seq_length=None, max_seq_length=None, min_gene_count=None, max_gene_count=None, nuc_completeness=None, has_proteins=None, proteins_complete=None, host_taxid=None, lab_passaged=None, geographic_region=None, geographic_location=None, submitter_country=None, min_collection_date=None, max_collection_date=None, annotated=None, source_database=None, min_release_date=None, max_release_date=None, min_mature_peptide_count=None, max_mature_peptide_count=None, min_protein_count=None, max_protein_count=None, max_ambiguous_chars=None)\n",
            "    Download a virus genome dataset from the NCBI Virus database (https://www.ncbi.nlm.nih.gov/labs/virus/).\n",
            "    \n",
            "    Args:\n",
            "    - virus                Virus taxon or accession, e.g. 'Norovirus' or 'coronaviridae' or\n",
            "                           '11320' (taxid of Influenza A virus) or 'NC_045512.2'\n",
            "                           If this input is a virus NCBI accession (e.g. 'NC_045512.2'), set accession = True.\n",
            "    - accession            True/False whether 'virus' is an accession. Default: False\n",
            "    - outfolder            Path to folder to save the requested data in, e.g. 'path/to/norovirus_folder'.\n",
            "                           Default: None (saves output into current working directory)\n",
            "    \n",
            "    Filters:\n",
            "    - host                 Host organism, e.g. 'homo sapiens'. Default: None\n",
            "    - min_seq_length       Min length of the returned sequences, e.g. 6252. Default: None\n",
            "    - max_seq_length       Max length of the returned sequences, e.g. 7500. Default: None\n",
            "    - min_gene_count       Min number of genes present in the virus genome, e.g. 1. Default: None\n",
            "    - max_gene_count       Max number of genes present in the virus genome, e.g. 40. Default: None\n",
            "    - nuc_completeness     Completeness status of the nucleotide sequence. Should be 'partial' or 'complete'. Default: None\n",
            "    - has_proteins         Str or list of proteins, genes, or segments that should be present in the sequence\n",
            "                           (based on sequence annotation provided by the submitter). Default: None\n",
            "                           Note: Set proteins_complete=True to only include sequences for which the proteins/genes/segments are marked 'complete'\n",
            "    - proteins_complete    Set to True if proteins/genes/segments in has_proteins should be marked as complete. Default: False\n",
            "    - host_taxid           NCBI Taxonomy ID of the host organism. Filters the results to only include viruses\n",
            "                           associated with hosts that fall within the specified TaxID. Default: None\n",
            "    - lab_passaged         True/False Indicates whether the virus sequence has been passaged in a laboratory setting.\n",
            "                           Default: None\n",
            "    - geographic_region    The geographic region where the virus was identified or isolated, e.g. 'Africa' or 'Europe'.\n",
            "                           Default: None\n",
            "    - geographic_location  The specific geographic location (e.g., country or area) where the virus was identified or\n",
            "                           isolated, e.g. 'South_Africa'. Default: None\n",
            "    - submitter_country    The country of the submitter of the virus sequence data. Filters results by the specified country,\n",
            "                           e.g. 'South_Africa' or 'Germany'. Default: None\n",
            "    - min_collection_date  The earliest collection date (in 'YYYY-MM-DD' format). Samples collected before this date will be\n",
            "                           excluded. Example: '2000-01-01'.  Default: None\n",
            "    - max_collection_date  The latest collection date (in 'YYYY-MM-DD' format). Samples collected after this date will be\n",
            "                           excluded. Example: '2014-12-04'. Default: None\n",
            "    - annotated            True/False Indicates whether the virus genome sequence is annotated. Default: None\n",
            "    - source_database      The source database from which the virus sequences originate, e.g. 'GenBank'.\n",
            "                           Default: None\n",
            "    - min_release_date     The earliest release date (in 'YYYY-MM-DD' format) of the virus sequence data. Sequences released\n",
            "                           before this date will be excluded. Default: None\n",
            "    - max_release_date     The latest release date (in 'YYYY-MM-DD' format) of the virus sequence data. Sequences released\n",
            "                           after this date will be excluded. Default: None\n",
            "    - min_mature_peptide_count    Min number of mature peptides present in the virus genome. Default: None\n",
            "    - max_mature_peptide_count    Max number of mature peptides present in the virus genome. Default: None\n",
            "    - min_protein_count    Min number of proteins present in the virus genome. Default: None\n",
            "    - max_protein_count    Max number of proteins present in the virus genome. Default: None\n",
            "    - max_ambiguous_chars  Max number of ambiguous nucleotide characters ('n' or 'N') allowed in the sequences, e.g. 50.\n",
            "                           Default: None\n",
            "    \n",
            "    Returns a fasta file containing the requested sequences and .csv and .jsonl files containing the associated metadata.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "help(gget.ncbi_virus)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5B5W5ioCy3AR"
      },
      "source": [
        "### Downloading virus genomes from NCBI Virus:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1Bvvj2mtn49X"
      },
      "source": [
        "This might take a minute depending on the internet connection and how busy the NCBI server is."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "vwDIiw3-yEtP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7342ae58-0542-462b-acbf-4bea4e9e01bf"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "New version of client (16.31.0) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets.\n",
            "INFO:gget.utils:229 sequences passed the provided filters.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 308 ms, sys: 31.4 ms, total: 340 ms\n",
            "Wall time: 2.52 s\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "gget.ncbi_virus(\n",
        "    virus = virus,\n",
        "    accession = accession,\n",
        "    host = host,\n",
        "    min_seq_length = min_seq_length,\n",
        "    max_seq_length = max_seq_length,\n",
        "    min_gene_count = min_gene_count,\n",
        "    max_gene_count = max_gene_count,\n",
        "    nuc_completeness = nuc_completeness,\n",
        "    has_proteins = has_proteins,\n",
        "    proteins_complete = proteins_complete,\n",
        "    host_taxid = host_taxid,\n",
        "    lab_passaged = lab_passaged,\n",
        "    geographic_region = geographic_region,\n",
        "    geographic_location = geographic_location,\n",
        "    submitter_country = submitter_country,\n",
        "    min_collection_date = min_collection_date,\n",
        "    max_collection_date = max_collection_date,\n",
        "    annotated = annotated,\n",
        "    source_database = source_database,\n",
        "    min_release_date = min_release_date,\n",
        "    max_release_date = max_release_date,\n",
        "    min_mature_peptide_count = min_mature_peptide_count,\n",
        "    max_mature_peptide_count = max_mature_peptide_count,\n",
        "    min_protein_count = min_protein_count,\n",
        "    max_protein_count = max_protein_count,\n",
        "    max_ambiguous_chars = max_ambiguous_chars\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "38O99mAsp3mP"
      },
      "source": [
        "### Merging sequencing and metadata files if additional file(s) were provided"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "cMlveCp8ompU"
      },
      "outputs": [],
      "source": [
        "ncbi_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences.fasta\"\n",
        "ncbi_metadata = f\"{'_'.join(str(virus).split(' '))}_metadata.csv\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EDLtDZ2wmGQv"
      },
      "source": [
        "If an additional fasta file with sequences was provided, adding these to the sequences and metadata to analyze:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "Z78AgbMOmFVb"
      },
      "outputs": [],
      "source": [
        "if fasta_file:\n",
        "  !pip install biopython\n",
        "  import pandas as pd\n",
        "  from Bio import SeqIO\n",
        "\n",
        "  # Combine sequence files\n",
        "  combined_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences_combined.fasta\"\n",
        "  !cat $ncbi_fasta_file $fasta_file > $combined_fasta_file\n",
        "  input_fasta_file = combined_fasta_file\n",
        "\n",
        "  # Combine metadata\n",
        "  combined_metadata_file = f\"{'_'.join(virus.split(' '))}_metadata_combined.csv\"\n",
        "  ncbi_metadata_df = pd.read_csv(ncbi_metadata)\n",
        "  if metadata_csv:\n",
        "    # Combine provided metadata and NCBI metadata csv files\n",
        "    user_metada_df = pd.read_csv(metadata_csv)\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  else:\n",
        "    # Extract sequence accessions from the provided FASTA file\n",
        "    headers = [record.id.split(\" \")[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "\n",
        "    # Create a metadata dataframe with the accessions from the FASTA file and the provided metadata\n",
        "    user_metada_df = pd.DataFrame(headers, columns=[\"Accession\"])\n",
        "    for key, value in metadata.items():\n",
        "      user_metada_df[key] = value\n",
        "\n",
        "    # Combine with NCBI metadata\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "else:\n",
        "  input_fasta_file = ncbi_fasta_file\n",
        "  metadata_file = ncbi_metadata"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UcxtXwMg_b4a"
      },
      "source": [
        "### Aligning the sequences to each other so they are all in the same frame:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l8VghGEXmHYa"
      },
      "source": [
        "Aligning all sequences in the faste file to each other so they are all in the same frame.\n",
        "\n",
        "Option 1: Using the MUSCLE algorithm (this works well for a few hundred sequences, but is too slow when dealing with a few thousand sequences)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "KVIBO6_16SZj"
      },
      "outputs": [],
      "source": [
        "# %%time\n",
        "# aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "# gget.muscle(input_fasta_file, super5=True, out=aligned_fasta_file)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Option 2: Using mafft"
      ],
      "metadata": {
        "id": "1zkJksf5j0YC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "#Installing MAFFT\n",
        "!apt-get install -qq -y mafft\n",
        "\n",
        "# Aligning sequences to each other using mafft\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "!mafft \\\n",
        "  --quiet \\\n",
        "  --auto \\\n",
        "  --thread 2 \\\n",
        "  $input_fasta_file > $aligned_fasta_file"
      ],
      "metadata": {
        "id": "kfa1lIuu1Z9n",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "03af3e34-7e09-4605-e926-d851f6009e75"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package fonts-lato.\n",
            "(Reading database ... 123629 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-lato_2.0-2.1_all.deb ...\n",
            "Unpacking fonts-lato (2.0-2.1) ...\n",
            "Selecting previously unselected package netbase.\n",
            "Preparing to unpack .../01-netbase_6.3_all.deb ...\n",
            "Unpacking netbase (6.3) ...\n",
            "Selecting previously unselected package libclone-perl.\n",
            "Preparing to unpack .../02-libclone-perl_0.45-1build3_amd64.deb ...\n",
            "Unpacking libclone-perl (0.45-1build3) ...\n",
            "Selecting previously unselected package libdata-dump-perl.\n",
            "Preparing to unpack .../03-libdata-dump-perl_1.25-1_all.deb ...\n",
            "Unpacking libdata-dump-perl (1.25-1) ...\n",
            "Selecting previously unselected package libencode-locale-perl.\n",
            "Preparing to unpack .../04-libencode-locale-perl_1.05-1.1_all.deb ...\n",
            "Unpacking libencode-locale-perl (1.05-1.1) ...\n",
            "Selecting previously unselected package libhttp-date-perl.\n",
            "Preparing to unpack .../05-libhttp-date-perl_6.05-1_all.deb ...\n",
            "Unpacking libhttp-date-perl (6.05-1) ...\n",
            "Selecting previously unselected package libfile-listing-perl.\n",
            "Preparing to unpack .../06-libfile-listing-perl_6.14-1_all.deb ...\n",
            "Unpacking libfile-listing-perl (6.14-1) ...\n",
            "Selecting previously unselected package libfont-afm-perl.\n",
            "Preparing to unpack .../07-libfont-afm-perl_1.20-3_all.deb ...\n",
            "Unpacking libfont-afm-perl (1.20-3) ...\n",
            "Selecting previously unselected package libhtml-tagset-perl.\n",
            "Preparing to unpack .../08-libhtml-tagset-perl_3.20-4_all.deb ...\n",
            "Unpacking libhtml-tagset-perl (3.20-4) ...\n",
            "Selecting previously unselected package liburi-perl.\n",
            "Preparing to unpack .../09-liburi-perl_5.10-1_all.deb ...\n",
            "Unpacking liburi-perl (5.10-1) ...\n",
            "Selecting previously unselected package libhtml-parser-perl:amd64.\n",
            "Preparing to unpack .../10-libhtml-parser-perl_3.76-1build2_amd64.deb ...\n",
            "Unpacking libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Selecting previously unselected package libio-html-perl.\n",
            "Preparing to unpack .../11-libio-html-perl_1.004-2_all.deb ...\n",
            "Unpacking libio-html-perl (1.004-2) ...\n",
            "Selecting previously unselected package liblwp-mediatypes-perl.\n",
            "Preparing to unpack .../12-liblwp-mediatypes-perl_6.04-1_all.deb ...\n",
            "Unpacking liblwp-mediatypes-perl (6.04-1) ...\n",
            "Selecting previously unselected package libhttp-message-perl.\n",
            "Preparing to unpack .../13-libhttp-message-perl_6.36-1_all.deb ...\n",
            "Unpacking libhttp-message-perl (6.36-1) ...\n",
            "Selecting previously unselected package libhtml-form-perl.\n",
            "Preparing to unpack .../14-libhtml-form-perl_6.07-1_all.deb ...\n",
            "Unpacking libhtml-form-perl (6.07-1) ...\n",
            "Selecting previously unselected package libhtml-tree-perl.\n",
            "Preparing to unpack .../15-libhtml-tree-perl_5.07-2_all.deb ...\n",
            "Unpacking libhtml-tree-perl (5.07-2) ...\n",
            "Selecting previously unselected package libhtml-format-perl.\n",
            "Preparing to unpack .../16-libhtml-format-perl_2.12-1.1_all.deb ...\n",
            "Unpacking libhtml-format-perl (2.12-1.1) ...\n",
            "Selecting previously unselected package libhttp-cookies-perl.\n",
            "Preparing to unpack .../17-libhttp-cookies-perl_6.10-1_all.deb ...\n",
            "Unpacking libhttp-cookies-perl (6.10-1) ...\n",
            "Selecting previously unselected package libhttp-daemon-perl.\n",
            "Preparing to unpack .../18-libhttp-daemon-perl_6.13-1ubuntu0.1_all.deb ...\n",
            "Unpacking libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libhttp-negotiate-perl.\n",
            "Preparing to unpack .../19-libhttp-negotiate-perl_6.01-1_all.deb ...\n",
            "Unpacking libhttp-negotiate-perl (6.01-1) ...\n",
            "Selecting previously unselected package perl-openssl-defaults:amd64.\n",
            "Preparing to unpack .../20-perl-openssl-defaults_5build2_amd64.deb ...\n",
            "Unpacking perl-openssl-defaults:amd64 (5build2) ...\n",
            "Selecting previously unselected package libnet-ssleay-perl:amd64.\n",
            "Preparing to unpack .../21-libnet-ssleay-perl_1.92-1build2_amd64.deb ...\n",
            "Unpacking libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Selecting previously unselected package libio-socket-ssl-perl.\n",
            "Preparing to unpack .../22-libio-socket-ssl-perl_2.074-2_all.deb ...\n",
            "Unpacking libio-socket-ssl-perl (2.074-2) ...\n",
            "Selecting previously unselected package libnet-http-perl.\n",
            "Preparing to unpack .../23-libnet-http-perl_6.22-1_all.deb ...\n",
            "Unpacking libnet-http-perl (6.22-1) ...\n",
            "Selecting previously unselected package libtry-tiny-perl.\n",
            "Preparing to unpack .../24-libtry-tiny-perl_0.31-1_all.deb ...\n",
            "Unpacking libtry-tiny-perl (0.31-1) ...\n",
            "Selecting previously unselected package libwww-robotrules-perl.\n",
            "Preparing to unpack .../25-libwww-robotrules-perl_6.02-1_all.deb ...\n",
            "Unpacking libwww-robotrules-perl (6.02-1) ...\n",
            "Selecting previously unselected package libwww-perl.\n",
            "Preparing to unpack .../26-libwww-perl_6.61-1_all.deb ...\n",
            "Unpacking libwww-perl (6.61-1) ...\n",
            "Selecting previously unselected package liblwp-protocol-https-perl.\n",
            "Preparing to unpack .../27-liblwp-protocol-https-perl_6.10-1_all.deb ...\n",
            "Unpacking liblwp-protocol-https-perl (6.10-1) ...\n",
            "Selecting previously unselected package libnet-smtp-ssl-perl.\n",
            "Preparing to unpack .../28-libnet-smtp-ssl-perl_1.04-1_all.deb ...\n",
            "Unpacking libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Selecting previously unselected package libmailtools-perl.\n",
            "Preparing to unpack .../29-libmailtools-perl_2.21-1_all.deb ...\n",
            "Unpacking libmailtools-perl (2.21-1) ...\n",
            "Selecting previously unselected package rubygems-integration.\n",
            "Preparing to unpack .../30-rubygems-integration_1.18_all.deb ...\n",
            "Unpacking rubygems-integration (1.18) ...\n",
            "Selecting previously unselected package ruby3.0.\n",
            "Preparing to unpack .../31-ruby3.0_3.0.2-7ubuntu2.7_amd64.deb ...\n",
            "Unpacking ruby3.0 (3.0.2-7ubuntu2.7) ...\n",
            "Selecting previously unselected package ruby-rubygems.\n",
            "Preparing to unpack .../32-ruby-rubygems_3.3.5-2_all.deb ...\n",
            "Unpacking ruby-rubygems (3.3.5-2) ...\n",
            "Selecting previously unselected package ruby.\n",
            "Preparing to unpack .../33-ruby_1%3a3.0~exp1_amd64.deb ...\n",
            "Unpacking ruby (1:3.0~exp1) ...\n",
            "Selecting previously unselected package rake.\n",
            "Preparing to unpack .../34-rake_13.0.6-2_all.deb ...\n",
            "Unpacking rake (13.0.6-2) ...\n",
            "Selecting previously unselected package ruby-net-telnet.\n",
            "Preparing to unpack .../35-ruby-net-telnet_0.1.1-2_all.deb ...\n",
            "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
            "Selecting previously unselected package ruby-webrick.\n",
            "Preparing to unpack .../36-ruby-webrick_1.7.0-3ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Selecting previously unselected package ruby-xmlrpc.\n",
            "Preparing to unpack .../37-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libruby3.0:amd64.\n",
            "Preparing to unpack .../38-libruby3.0_3.0.2-7ubuntu2.7_amd64.deb ...\n",
            "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.7) ...\n",
            "Selecting previously unselected package lynx-common.\n",
            "Preparing to unpack .../39-lynx-common_2.9.0dev.10-1_all.deb ...\n",
            "Unpacking lynx-common (2.9.0dev.10-1) ...\n",
            "Selecting previously unselected package mafft.\n",
            "Preparing to unpack .../40-mafft_7.490-1_amd64.deb ...\n",
            "Unpacking mafft (7.490-1) ...\n",
            "Selecting previously unselected package mailcap.\n",
            "Preparing to unpack .../41-mailcap_3.70+nmu1ubuntu1_all.deb ...\n",
            "Unpacking mailcap (3.70+nmu1ubuntu1) ...\n",
            "Selecting previously unselected package mime-support.\n",
            "Preparing to unpack .../42-mime-support_3.66_all.deb ...\n",
            "Unpacking mime-support (3.66) ...\n",
            "Selecting previously unselected package libauthen-sasl-perl.\n",
            "Preparing to unpack .../43-libauthen-sasl-perl_2.1600-1.1_all.deb ...\n",
            "Unpacking libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Selecting previously unselected package lynx.\n",
            "Preparing to unpack .../44-lynx_2.9.0dev.10-1_amd64.deb ...\n",
            "Unpacking lynx (2.9.0dev.10-1) ...\n",
            "Setting up libhttp-date-perl (6.05-1) ...\n",
            "Setting up fonts-lato (2.0-2.1) ...\n",
            "Setting up libfile-listing-perl (6.14-1) ...\n",
            "Setting up libfont-afm-perl (1.20-3) ...\n",
            "Setting up mafft (7.490-1) ...\n",
            "Setting up libclone-perl (0.45-1build3) ...\n",
            "Setting up libhtml-tagset-perl (3.20-4) ...\n",
            "Setting up libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Setting up liblwp-mediatypes-perl (6.04-1) ...\n",
            "Setting up libtry-tiny-perl (0.31-1) ...\n",
            "Setting up perl-openssl-defaults:amd64 (5build2) ...\n",
            "Setting up libencode-locale-perl (1.05-1.1) ...\n",
            "Setting up rubygems-integration (1.18) ...\n",
            "Setting up libdata-dump-perl (1.25-1) ...\n",
            "Setting up ruby-net-telnet (0.1.1-2) ...\n",
            "Setting up libio-html-perl (1.004-2) ...\n",
            "Setting up lynx-common (2.9.0dev.10-1) ...\n",
            "Setting up ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Setting up netbase (6.3) ...\n",
            "Setting up lynx (2.9.0dev.10-1) ...\n",
            "update-alternatives: using /usr/bin/lynx to provide /usr/bin/www-browser (www-browser) in auto mode\n",
            "Setting up mailcap (3.70+nmu1ubuntu1) ...\n",
            "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Setting up liburi-perl (5.10-1) ...\n",
            "Setting up libhttp-message-perl (6.36-1) ...\n",
            "Setting up libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Setting up libhttp-negotiate-perl (6.01-1) ...\n",
            "Setting up mime-support (3.66) ...\n",
            "Setting up libhttp-cookies-perl (6.10-1) ...\n",
            "Setting up libnet-http-perl (6.22-1) ...\n",
            "Setting up libwww-robotrules-perl (6.02-1) ...\n",
            "Setting up libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Setting up libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Setting up libio-socket-ssl-perl (2.074-2) ...\n",
            "Setting up libhtml-form-perl (6.07-1) ...\n",
            "Setting up libhtml-tree-perl (5.07-2) ...\n",
            "Setting up libhtml-format-perl (2.12-1.1) ...\n",
            "Setting up libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Setting up libmailtools-perl (2.21-1) ...\n",
            "Setting up libwww-perl (6.61-1) ...\n",
            "Setting up ruby3.0 (3.0.2-7ubuntu2.7) ...\n",
            "Setting up liblwp-protocol-https-perl (6.10-1) ...\n",
            "Setting up ruby (1:3.0~exp1) ...\n",
            "Setting up rake (13.0.6-2) ...\n",
            "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.7) ...\n",
            "Setting up ruby-rubygems (3.3.5-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "CPU times: user 608 ms, sys: 84.4 ms, total: 693 ms\n",
            "Wall time: 1min 9s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# gget.mafft(input_fasta_file, out=aligned_fasta_file)"
      ],
      "metadata": {
        "id": "RHQpFzcQ1fPK"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z3uqWWpJ_fA_"
      },
      "source": [
        "### Running Delphy:"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Adjust the headers in the aligned fasta file to match header format required by Delphy (accession|YYYY-MM-DD):"
      ],
      "metadata": {
        "id": "Hyf2y0YOpIRE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "import pandas as pd\n",
        "import re\n",
        "from datetime import datetime"
      ],
      "metadata": {
        "id": "PUH80U2BpBzl"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Reformat collection date\n",
        "default_day = '01'\n",
        "default_month = '01'\n",
        "def extract_and_format_date(date_string):\n",
        "    # Define regular expressions for various date formats\n",
        "    year_only = re.compile(r'(?P<year>\\d{4})')\n",
        "    year_month = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})')\n",
        "    full_date = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})[-/.](?P<day>\\d{1,2})')\n",
        "\n",
        "    # Try to match the full date first\n",
        "    match = full_date.search(date_string)\n",
        "    if match:\n",
        "        year = match.group('year')\n",
        "        month = match.group('month').zfill(2)\n",
        "        day = match.group('day').zfill(2)\n",
        "    else:\n",
        "        # Try to match year and month\n",
        "        match = year_month.search(date_string)\n",
        "        if match:\n",
        "            year = match.group('year')\n",
        "            month = match.group('month').zfill(2)\n",
        "            day = default_day\n",
        "        else:\n",
        "            # Try to match only the year\n",
        "            match = year_only.search(date_string)\n",
        "            if match:\n",
        "                year = match.group('year')\n",
        "                month = default_month\n",
        "                day = default_day\n",
        "            else:\n",
        "                # If no match, return None\n",
        "                return None\n",
        "\n",
        "    # Format the extracted date into YYYY-MM-DD\n",
        "    formatted_date = f\"{year}-{month}-{day}\"\n",
        "\n",
        "    try:\n",
        "        # Validate date by trying to convert it to a datetime object\n",
        "        datetime.strptime(formatted_date, '%Y-%m-%d')\n",
        "    except ValueError:\n",
        "        return None  # Return None if the date is invalid\n",
        "\n",
        "    return formatted_date"
      ],
      "metadata": {
        "id": "LzhpfpACozJs"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def update_fasta_headers(fasta_file, csv_file, output_fasta):\n",
        "    # Read the CSV file into a pandas DataFrame\n",
        "    df = pd.read_csv(csv_file)\n",
        "\n",
        "    # Create a dictionary from the DataFrame to map accession to date\n",
        "    accession_to_date = pd.Series(df['Collection Date'].values, index=df['accession']).to_dict()\n",
        "\n",
        "    # Open the input fasta file and output fasta file\n",
        "    with open(fasta_file) as fasta_input, open(output_fasta, 'w') as fasta_output:\n",
        "        # Iterate through each sequence record in the fasta file\n",
        "        for record in SeqIO.parse(fasta_input, 'fasta'):\n",
        "            accession = record.id\n",
        "\n",
        "            # Check if the accession is in the pandas dictionary and has a non-NaN date\n",
        "            if accession in accession_to_date and pd.notna(accession_to_date[accession]):\n",
        "                date = accession_to_date[accession]\n",
        "\n",
        "                # Format the date if necessary\n",
        "                formatted_date = extract_and_format_date(date)\n",
        "\n",
        "                if formatted_date is None:\n",
        "                  # Skip the entry if date is NaN or accession not found\n",
        "                  print(f\"Skipping accession {accession} due to unrecognized date format: '{date}'\")\n",
        "                  continue\n",
        "\n",
        "                # Update the seq header\n",
        "                record.id = f\"{accession}|{formatted_date}\"\n",
        "                record.description = ''  # Remove the original description to avoid duplication\n",
        "            else:\n",
        "                # Skip the entry if date is NaN or accession not found\n",
        "                print(f\"Skipping accession {accession} due to missing or NaN date.\")\n",
        "                continue\n",
        "\n",
        "            # Write the updated record to the output fasta file\n",
        "            SeqIO.write(record, fasta_output, 'fasta')\n",
        "\n",
        "aligned_fasta_file_clean = f\"{'_'.join(str(virus).split(' '))}_aligned_headers_adjusted.afa\"\n",
        "update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3Q5aX8uD1tPm",
        "outputId": "fd0cccb7-6f7b-4333-ad36-d031438fd700"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Skipping accession AY947539.1 due to missing or NaN date.\n",
            "Skipping accession KJ160504.1 due to missing or NaN date.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Run Delphy:"
      ],
      "metadata": {
        "id": "U3KdqxD3pL_a"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Download delphy binary\n",
        "!wget https://github.com/broadinstitute/delphy/releases/download/0.9995/delphy-ubuntu-x86_64\n",
        "\n",
        "# Give permissions\n",
        "!chmod u+x ./delphy-ubuntu-x86_64"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UKhk89521Amh",
        "outputId": "2ba798ba-237f-4ae9-be82-016e87de2316"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-10-17 22:47:20--  https://github.com/broadinstitute/delphy/releases/download/0.9995/delphy-ubuntu-x86_64\n",
            "Resolving github.com (github.com)... 140.82.112.4\n",
            "Connecting to github.com (github.com)|140.82.112.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/804287442/5e0b7be1-42ce-451f-a096-167441c1a12a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241017%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241017T224720Z&X-Amz-Expires=300&X-Amz-Signature=e8103ebec8959e4beeddf4d3886b02e79814d1bad90859e971562656d4247127&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Ddelphy-ubuntu-x86_64&response-content-type=application%2Foctet-stream [following]\n",
            "--2024-10-17 22:47:20--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/804287442/5e0b7be1-42ce-451f-a096-167441c1a12a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241017%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241017T224720Z&X-Amz-Expires=300&X-Amz-Signature=e8103ebec8959e4beeddf4d3886b02e79814d1bad90859e971562656d4247127&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Ddelphy-ubuntu-x86_64&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 2165232 (2.1M) [application/octet-stream]\n",
            "Saving to: ‘delphy-ubuntu-x86_64’\n",
            "\n",
            "delphy-ubuntu-x86_6 100%[===================>]   2.06M  --.-KB/s    in 0.07s   \n",
            "\n",
            "2024-10-17 22:47:20 (29.9 MB/s) - ‘delphy-ubuntu-x86_64’ saved [2165232/2165232]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Run delphy\n",
        "threads = 2\n",
        "\n",
        "!./delphy-ubuntu-x86_64 \\\n",
        "  --v0-threads $threads \\\n",
        "  --v0-in-fasta $aligned_fasta_file_clean"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zvcsj4Rk01fA",
        "outputId": "31480a55-a6c2-4cc5-8648-c3d63f5385cb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Seed: 2556349609\n",
            "Reading fasta file 11070_aligned_headers_adjusted.afa\n",
            "Building rough initial tree...\n",
            "Warning: ambiguous state Y at site 123 changed to N\n",
            "Warning: ambiguous state Y at site 2513 changed to N\n",
            "Warning: ambiguous state Y at site 7214 changed to N\n",
            "Warning: ambiguous state Y at site 7291 changed to N\n",
            "Warning: ambiguous state S at site 274 changed to N\n",
            "Warning: ambiguous state Y at site 367 changed to N\n",
            "Warning: ambiguous state Y at site 637 changed to N\n",
            "Warning: ambiguous state M at site 1757 changed to N\n",
            "Warning: ambiguous state S at site 2825 changed to N\n",
            "Warning: ambiguous state Y at site 3065 changed to N\n",
            "Warning: ambiguous state R at site 5834 changed to N\n",
            "Warning: ambiguous state R at site 9611 changed to N\n",
            "Warning: ambiguous state R at site 2249 changed to N\n",
            "Warning: ambiguous state S at site 3236 changed to N\n",
            "Warning: ambiguous state S at site 4118 changed to N\n",
            "Warning: ambiguous state R at site 4208 changed to N\n",
            "Warning: ambiguous state W at site 5786 changed to N\n",
            "Warning: ambiguous state M at site 6536 changed to N\n",
            "Warning: ambiguous state R at site 9146 changed to N\n",
            "Warning: ambiguous state M at site 1985 changed to N\n",
            "Warning: ambiguous state W at site 2623 changed to N\n",
            "Warning: ambiguous state W at site 2629 changed to N\n",
            "Warning: ambiguous state K at site 9596 changed to N\n",
            "Warning: ambiguous state W at site 530 changed to N\n",
            "Warning: ambiguous state K at site 4222 changed to N\n",
            "Warning: ambiguous state R at site 4223 changed to N\n",
            "Warning: ambiguous state W at site 4224 changed to N\n",
            "Warning: ambiguous state S at site 4225 changed to N\n",
            "Warning: ambiguous state Y at site 7040 changed to N\n",
            "Warning: ambiguous state Y at site 1009 changed to N\n",
            "Warning: ambiguous state Y at site 8146 changed to N\n",
            "Warning: ambiguous state M at site 8536 changed to N\n",
            "Warning: ambiguous state Y at site 3972 changed to N\n",
            "Warning: ambiguous state Y at site 8146 changed to N\n",
            "Warning: ambiguous state S at site 8552 changed to N\n",
            "Warning: ambiguous state W at site 8588 changed to N\n",
            "Warning: ambiguous state R at site 5276 changed to N\n",
            "Warning: ambiguous state R at site 5828 changed to N\n",
            "Warning: ambiguous state R at site 5831 changed to N\n",
            "Warning: ambiguous state M at site 2732 changed to N\n",
            "Warning: ambiguous state Y at site 4503 changed to N\n",
            "Warning: ambiguous state M at site 4536 changed to N\n",
            "Warning: ambiguous state M at site 4564 changed to N\n",
            "Warning: ambiguous state R at site 7892 changed to N\n",
            "Warning: ambiguous state R at site 8169 changed to N\n",
            "Warning: ambiguous state Y at site 9434 changed to N\n",
            "Warning: ambiguous state W at site 7086 changed to N\n",
            "Warning: ambiguous state Y at site 7675 changed to N\n",
            "Warning: ambiguous state D at site 7752 changed to N\n",
            "Warning: ambiguous state R at site 10865 changed to N\n",
            "Warning: ambiguous state Y at site 310 changed to N\n",
            "Warning: ambiguous state K at site 2111 changed to N\n",
            "Warning: ambiguous state R at site 2348 changed to N\n",
            "Warning: ambiguous state W at site 2817 changed to N\n",
            "Warning: ambiguous state W at site 3383 changed to N\n",
            "Warning: ambiguous state R at site 4734 changed to N\n",
            "Warning: ambiguous state Y at site 5615 changed to N\n",
            "Warning: ambiguous state R at site 5774 changed to N\n",
            "Warning: ambiguous state Y at site 6296 changed to N\n",
            "Warning: ambiguous state Y at site 6305 changed to N\n",
            "Warning: ambiguous state Y at site 6443 changed to N\n",
            "Warning: ambiguous state Y at site 6839 changed to N\n",
            "Warning: ambiguous state Y at site 8432 changed to N\n",
            "Warning: ambiguous state Y at site 10121 changed to N\n",
            "Warning: ambiguous state R at site 10322 changed to N\n",
            "Warning: ambiguous state Y at site 10375 changed to N\n",
            "Warning: ambiguous state M at site 10583 changed to N\n",
            "Warning: ambiguous state Y at site 10607 changed to N\n",
            "Warning: ambiguous state Y at site 10659 changed to N\n",
            "Warning: ambiguous state R at site 10830 changed to N\n",
            "Warning: ambiguous state R at site 10860 changed to N\n",
            "Warning: ambiguous state R at site 10867 changed to N\n",
            "Warning: ambiguous state Y at site 310 changed to N\n",
            "Warning: ambiguous state R at site 1060 changed to N\n",
            "Warning: ambiguous state Y at site 1072 changed to N\n",
            "Warning: ambiguous state R at site 1366 changed to N\n",
            "Warning: ambiguous state R at site 1405 changed to N\n",
            "Warning: ambiguous state Y at site 1849 changed to N\n",
            "Warning: ambiguous state Y at site 2762 changed to N\n",
            "Warning: ambiguous state R at site 4734 changed to N\n",
            "Warning: ambiguous state Y at site 5615 changed to N\n",
            "Warning: ambiguous state Y at site 6443 changed to N\n",
            "Warning: ambiguous state Y at site 6459 changed to N\n",
            "Warning: ambiguous state Y at site 6648 changed to N\n",
            "Warning: ambiguous state Y at site 6761 changed to N\n",
            "Warning: ambiguous state R at site 6779 changed to N\n",
            "Warning: ambiguous state Y at site 6839 changed to N\n",
            "Warning: ambiguous state Y at site 7067 changed to N\n",
            "Warning: ambiguous state Y at site 7079 changed to N\n",
            "Warning: ambiguous state R at site 7163 changed to N\n",
            "Warning: ambiguous state R at site 7235 changed to N\n",
            "Warning: ambiguous state Y at site 7241 changed to N\n",
            "Warning: ambiguous state Y at site 7346 changed to N\n",
            "Warning: ambiguous state R at site 7376 changed to N\n",
            "Warning: ambiguous state Y at site 7382 changed to N\n",
            "Warning: ambiguous state Y at site 7480 changed to N\n",
            "Warning: ambiguous state R at site 7628 changed to N\n",
            "Warning: ambiguous state Y at site 7791 changed to N\n",
            "Warning: ambiguous state Y at site 7973 changed to N\n",
            "Warning: ambiguous state Y at site 8033 changed to N\n",
            "Warning: ambiguous state R at site 8165 changed to N\n",
            "Warning: ambiguous state Y at site 8297 changed to N\n",
            "Warning: ambiguous state Y at site 8432 changed to N\n",
            "Warning: ambiguous state Y at site 9815 changed to N\n",
            "Warning: ambiguous state Y at site 10121 changed to N\n",
            "Warning: ambiguous state R at site 10214 changed to N\n",
            "Warning: ambiguous state Y at site 10375 changed to N\n",
            "Warning: ambiguous state M at site 10583 changed to N\n",
            "Warning: ambiguous state Y at site 10607 changed to N\n",
            "Attaching tip 2 to Candidate_region{branch=0, mut_idx=78, -12421.3<t<=-12418.9, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 3 to Candidate_region{branch=227, mut_idx=0, -1.79769e+308<t<=-13446, min_muts=102, W_over_Wmax=0}\n",
            "Attaching tip 4 to Candidate_region{branch=3, mut_idx=102, -4876.88<t<=-4748, min_muts=88, W_over_Wmax=0}\n",
            "Attaching tip 5 to Candidate_region{branch=4, mut_idx=7, -4792.17<t<=-4790.62, min_muts=72, W_over_Wmax=0}\n",
            "Attaching tip 6 to Candidate_region{branch=2, mut_idx=25, -8057.7<t<=-8035, min_muts=41, W_over_Wmax=0}\n",
            "Attaching tip 7 to Candidate_region{branch=228, mut_idx=8, -13320.9<t<=-13319, min_muts=216, W_over_Wmax=0}\n",
            "Attaching tip 8 to Candidate_region{branch=5, mut_idx=70, -4749.05<t<=-4748.81, min_muts=52, W_over_Wmax=0}\n",
            "Attaching tip 9 to Candidate_region{branch=8, mut_idx=50, -4748.06<t<=-4748.03, min_muts=48, W_over_Wmax=0}\n",
            "Attaching tip 10 to Candidate_region{branch=7, mut_idx=1, -13312.1<t<=-13310, min_muts=159, W_over_Wmax=0}\n",
            "Attaching tip 11 to Candidate_region{branch=233, mut_idx=8, -13320.9<t<=-13319.1, min_muts=77, W_over_Wmax=0}\n",
            "Attaching tip 12 to Candidate_region{branch=11, mut_idx=3, -13116<t<=-13009.6, min_muts=72, W_over_Wmax=0}\n",
            "Attaching tip 13 to Candidate_region{branch=12, mut_idx=1, -12956.4<t<=-12885.2, min_muts=68, W_over_Wmax=0}\n",
            "Attaching tip 14 to Candidate_region{branch=11, mut_idx=74, -8057.89<t<=-8035, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 15 to Candidate_region{branch=14, mut_idx=0, -8054.22<t<=-8052.57, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 16 to Candidate_region{branch=240, mut_idx=73, -8059.17<t<=-8057.89, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 17 to Candidate_region{branch=11, mut_idx=0, -8054.22<t<=-8035, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 18 to Candidate_region{branch=17, mut_idx=13, -7670.28<t<=-7670, min_muts=22, W_over_Wmax=0}\n",
            "Attaching tip 19 to Candidate_region{branch=13, mut_idx=67, -8103<t<=-8047, min_muts=37, W_over_Wmax=0}\n",
            "Attaching tip 20 to Candidate_region{branch=230, mut_idx=6, -12777.6<t<=-12749, min_muts=81, W_over_Wmax=0}\n",
            "Attaching tip 21 to Candidate_region{branch=10, mut_idx=159, -9523.76<t<=-9496, min_muts=96, W_over_Wmax=0}\n",
            "Attaching tip 22 to Candidate_region{branch=21, mut_idx=96, -9132.91<t<=-9131, min_muts=35, W_over_Wmax=0}\n",
            "Attaching tip 23 to Candidate_region{branch=232, mut_idx=15, -9892.88<t<=-9667.72, min_muts=19, W_over_Wmax=0}\n",
            "Attaching tip 24 to Candidate_region{branch=1, mut_idx=87, -8046.8<t<=-8035, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 25 to Candidate_region{branch=20, mut_idx=74, -7318.19<t<=-7305, min_muts=48, W_over_Wmax=0}\n",
            "Attaching tip 26 to Candidate_region{branch=8, mut_idx=2, -4748<t<=-4748, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 27 to Candidate_region{branch=4, mut_idx=81, -4748.08<t<=-4748, min_muts=20, W_over_Wmax=0}\n",
            "Attaching tip 28 to Candidate_region{branch=27, mut_idx=20, -4748.01<t<=-4748, min_muts=22, W_over_Wmax=0}\n",
            "Attaching tip 29 to Candidate_region{branch=3, mut_idx=0, -4797.33<t<=-4748, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 30 to Candidate_region{branch=26, mut_idx=16, -4748<t<=-4748, min_muts=34, W_over_Wmax=0}\n",
            "Attaching tip 31 to Candidate_region{branch=30, mut_idx=5, -4748<t<=-4748, min_muts=43, W_over_Wmax=0}\n",
            "Attaching tip 32 to Candidate_region{branch=29, mut_idx=1, -4772.33<t<=-4759.14, min_muts=58, W_over_Wmax=0}\n",
            "Attaching tip 33 to Candidate_region{branch=32, mut_idx=58, -4748.37<t<=-4748, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 34 to Candidate_region{branch=28, mut_idx=17, -4748<t<=-4748, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 35 to Candidate_region{branch=258, mut_idx=1, -4772.33<t<=-4765.61, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 36 to Candidate_region{branch=29, mut_idx=3, -4748.21<t<=-4748, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 37 to Candidate_region{branch=36, mut_idx=2, -4748.02<t<=-4748, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 38 to Candidate_region{branch=5, mut_idx=2, -4748.15<t<=-4748, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 39 to Candidate_region{branch=23, mut_idx=19, -8776.31<t<=-8766, min_muts=30, W_over_Wmax=0}\n",
            "Attaching tip 40 to Candidate_region{branch=39, mut_idx=30, -8092.69<t<=-8035, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 41 to Candidate_region{branch=266, mut_idx=0, -8775.23<t<=-8736.41, min_muts=21, W_over_Wmax=0}\n",
            "Attaching tip 42 to Candidate_region{branch=267, mut_idx=0, -8775.23<t<=-8740.09, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 43 to Candidate_region{branch=6, mut_idx=41, -8035.08<t<=-8035, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 44 to Candidate_region{branch=0, mut_idx=1, -12418.9<t<=-12418, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 45 to Candidate_region{branch=266, mut_idx=2, -8721.26<t<=-8708.31, min_muts=55, W_over_Wmax=0}\n",
            "Attaching tip 46 to Candidate_region{branch=31, mut_idx=40, -4748<t<=-4748, min_muts=47, W_over_Wmax=0}\n",
            "Attaching tip 47 to Candidate_region{branch=265, mut_idx=6, -9500.59<t<=-9496, min_muts=24, W_over_Wmax=0}\n",
            "Attaching tip 48 to Candidate_region{branch=248, mut_idx=3, -9507.7<t<=-9496, min_muts=87, W_over_Wmax=0}\n",
            "Attaching tip 49 to Candidate_region{branch=48, mut_idx=87, -9496.05<t<=-9496, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 50 to Candidate_region{branch=47, mut_idx=24, -9496.04<t<=-9496, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 51 to Candidate_region{branch=268, mut_idx=0, -8775.23<t<=-8766, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 52 to Candidate_region{branch=50, mut_idx=14, -8829.8<t<=-8798.39, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 53 to Candidate_region{branch=52, mut_idx=10, -8766.06<t<=-8766, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 54 to Candidate_region{branch=51, mut_idx=4, -8768.05<t<=-8767.56, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 55 to Candidate_region{branch=49, mut_idx=1, -9470.63<t<=-9460.79, min_muts=24, W_over_Wmax=0}\n",
            "Attaching tip 56 to Candidate_region{branch=7, mut_idx=17, -12946.8<t<=-12921.6, min_muts=605, W_over_Wmax=0}\n",
            "Attaching tip 57 to Candidate_region{branch=278, mut_idx=2, -9412.88<t<=-9286.84, min_muts=110, W_over_Wmax=0}\n",
            "Attaching tip 58 to Candidate_region{branch=57, mut_idx=102, -7064.08<t<=-7056.15, min_muts=63, W_over_Wmax=0}\n",
            "Attaching tip 59 to Candidate_region{branch=58, mut_idx=63, -6939.83<t<=-6939, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 60 to Candidate_region{branch=285, mut_idx=63, -6939.83<t<=-6939.09, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 61 to Candidate_region{branch=20, mut_idx=1, -7284.02<t<=-7257.57, min_muts=65, W_over_Wmax=0}\n",
            "Attaching tip 62 to Candidate_region{branch=61, mut_idx=65, -5877.47<t<=-5844, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 63 to Candidate_region{branch=230, mut_idx=85, -5544.49<t<=-5478, min_muts=45, W_over_Wmax=0}\n",
            "Attaching tip 64 to Candidate_region{branch=229, mut_idx=0, -1.79769e+308<t<=-23376, min_muts=570, W_over_Wmax=0}\n",
            "Attaching tip 65 to Candidate_region{branch=263, mut_idx=2, -4748.02<t<=-4748, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 66 to Candidate_region{branch=262, mut_idx=3, -4748.21<t<=-4748.05, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 67 to Candidate_region{branch=37, mut_idx=0, -4748<t<=-4748, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 68 to Candidate_region{branch=292, mut_idx=1, -4759.14<t<=-4753.17, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 69 to Candidate_region{branch=290, mut_idx=0, -1.79769e+308<t<=-30787, min_muts=117, W_over_Wmax=0}\n",
            "Attaching tip 70 to Candidate_region{branch=9, mut_idx=48, -4748<t<=-4748, min_muts=17, W_over_Wmax=0}\n",
            "Attaching tip 71 to Candidate_region{branch=70, mut_idx=14, -4748<t<=-4748, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 72 to Candidate_region{branch=272, mut_idx=18, -4748<t<=-4748, min_muts=50, W_over_Wmax=0}\n",
            "Attaching tip 73 to Candidate_region{branch=33, mut_idx=12, -4748<t<=-4748, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 74 to Candidate_region{branch=63, mut_idx=45, -5478.43<t<=-5478, min_muts=39, W_over_Wmax=0}\n",
            "Attaching tip 75 to Candidate_region{branch=7, mut_idx=168, -9143.4<t<=-9131, min_muts=107, W_over_Wmax=0}\n",
            "Attaching tip 76 to Candidate_region{branch=69, mut_idx=114, -14445.4<t<=-14424, min_muts=127, W_over_Wmax=0}\n",
            "Attaching tip 77 to Candidate_region{branch=64, mut_idx=568, -23398.8<t<=-23395.7, min_muts=500, W_over_Wmax=0}\n",
            "Attaching tip 78 to Candidate_region{branch=56, mut_idx=605, -5481.4<t<=-5478, min_muts=391, W_over_Wmax=0}\n",
            "Attaching tip 79 to Candidate_region{branch=69, mut_idx=3, -14010.3<t<=-13879, min_muts=31, W_over_Wmax=0}\n",
            "Attaching tip 80 to Candidate_region{branch=74, mut_idx=36, -4849.81<t<=-4837.44, min_muts=60, W_over_Wmax=0}\n",
            "Attaching tip 81 to Candidate_region{branch=77, mut_idx=8, -23379.5<t<=-23368.6, min_muts=577, W_over_Wmax=0}\n",
            "Attaching tip 82 to Candidate_region{branch=81, mut_idx=562, -4764.87<t<=-4748, min_muts=52, W_over_Wmax=0}\n",
            "Attaching tip 83 to Candidate_region{branch=82, mut_idx=52, -4748.05<t<=-4748, min_muts=138, W_over_Wmax=0}\n",
            "Attaching tip 84 to Candidate_region{branch=306, mut_idx=0, -5478.17<t<=-5463.92, min_muts=47, W_over_Wmax=0}\n",
            "Attaching tip 85 to Candidate_region{branch=13, mut_idx=0, -8047.42<t<=-8047, min_muts=95, W_over_Wmax=0}\n",
            "Attaching tip 86 to Candidate_region{branch=80, mut_idx=60, -3456.47<t<=-3451, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 87 to Candidate_region{branch=30, mut_idx=28, -4748<t<=-4748, min_muts=48, W_over_Wmax=0}\n",
            "Attaching tip 88 to Candidate_region{branch=87, mut_idx=48, -3552<t<=-3455, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 89 to Candidate_region{branch=88, mut_idx=14, -3459.35<t<=-3454.6, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 90 to Candidate_region{branch=314, mut_idx=13, -4319.85<t<=-4314.02, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 91 to Candidate_region{branch=53, mut_idx=8, -8766.01<t<=-8766, min_muts=78, W_over_Wmax=0}\n",
            "Attaching tip 92 to Candidate_region{branch=91, mut_idx=78, -3428.52<t<=-3339, min_muts=26, W_over_Wmax=0}\n",
            "Attaching tip 93 to Candidate_region{branch=92, mut_idx=25, -3280.57<t<=-3277, min_muts=38, W_over_Wmax=0}\n",
            "Attaching tip 94 to Candidate_region{branch=93, mut_idx=23, -3277.21<t<=-3277.19, min_muts=19, W_over_Wmax=0}\n",
            "Attaching tip 95 to Candidate_region{branch=306, mut_idx=3, -5429.53<t<=-5409.6, min_muts=55, W_over_Wmax=0}\n",
            "Attaching tip 96 to Candidate_region{branch=95, mut_idx=54, -3373.6<t<=-3275.44, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 97 to Candidate_region{branch=90, mut_idx=25, -3474.72<t<=-3417, min_muts=19, W_over_Wmax=0}\n",
            "Attaching tip 98 to Candidate_region{branch=323, mut_idx=25, -3474.72<t<=-3436.39, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 99 to Candidate_region{branch=96, mut_idx=11, -3285.27<t<=-3279.5, min_muts=36, W_over_Wmax=0}\n",
            "Attaching tip 100 to Candidate_region{branch=99, mut_idx=42, -3260.13<t<=-3259.2, min_muts=44, W_over_Wmax=0}\n",
            "Attaching tip 101 to Candidate_region{branch=95, mut_idx=1, -3275.44<t<=-3275, min_muts=40, W_over_Wmax=0}\n",
            "Attaching tip 102 to Candidate_region{branch=83, mut_idx=1, -4742.41<t<=-4739.41, min_muts=277, W_over_Wmax=0}\n",
            "Attaching tip 103 to Candidate_region{branch=78, mut_idx=343, -3657.94<t<=-3652, min_muts=55, W_over_Wmax=0}\n",
            "Attaching tip 104 to Candidate_region{branch=77, mut_idx=492, -21554.2<t<=-21549, min_muts=223, W_over_Wmax=0}\n",
            "Attaching tip 105 to Candidate_region{branch=307, mut_idx=2, -23392.1<t<=-23389.4, min_muts=470, W_over_Wmax=0}\n",
            "Attaching tip 106 to Candidate_region{branch=105, mut_idx=446, -15533.1<t<=-15528.9, min_muts=473, W_over_Wmax=0}\n",
            "Attaching tip 107 to Candidate_region{branch=103, mut_idx=55, -3652.03<t<=-3652, min_muts=233, W_over_Wmax=0}\n",
            "Attaching tip 108 to Candidate_region{branch=107, mut_idx=233, -3318.13<t<=-3318, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 109 to Candidate_region{branch=56, mut_idx=0, -5480.05<t<=-5478, min_muts=133, W_over_Wmax=0}\n",
            "Attaching tip 110 to Candidate_region{branch=329, mut_idx=167, -4603.75<t<=-4599, min_muts=248, W_over_Wmax=0}\n",
            "Attaching tip 111 to Candidate_region{branch=110, mut_idx=248, -4599.01<t<=-4599, min_muts=237, W_over_Wmax=0}\n",
            "Attaching tip 112 to Candidate_region{branch=111, mut_idx=0, -4599<t<=-4598.87, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 113 to Candidate_region{branch=111, mut_idx=238, -4559.06<t<=-4559, min_muts=80, W_over_Wmax=0}\n",
            "Attaching tip 114 to Candidate_region{branch=112, mut_idx=0, -4598.99<t<=-4598.74, min_muts=63, W_over_Wmax=0}\n",
            "Attaching tip 115 to Candidate_region{branch=114, mut_idx=0, -4598.79<t<=-4597.91, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 116 to Candidate_region{branch=106, mut_idx=474, -4023.13<t<=-4017, min_muts=161, W_over_Wmax=0}\n",
            "Attaching tip 117 to Candidate_region{branch=308, mut_idx=503, -6623.9<t<=-6574, min_muts=326, W_over_Wmax=0}\n",
            "Attaching tip 118 to Candidate_region{branch=333, mut_idx=44, -3652.85<t<=-3652.55, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 119 to Candidate_region{branch=86, mut_idx=10, -3420.25<t<=-3412.53, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 120 to Candidate_region{branch=119, mut_idx=0, -3415.54<t<=-3408.18, min_muts=23, W_over_Wmax=0}\n",
            "Attaching tip 121 to Candidate_region{branch=120, mut_idx=23, -2818.45<t<=-2814, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 122 to Candidate_region{branch=347, mut_idx=20, -2840.26<t<=-2836, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 123 to Candidate_region{branch=119, mut_idx=20, -3115.44<t<=-3031.65, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 124 to Candidate_region{branch=123, mut_idx=13, -2942.26<t<=-2928.79, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 125 to Candidate_region{branch=122, mut_idx=9, -2835.38<t<=-2835.34, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 126 to Candidate_region{branch=124, mut_idx=5, -2907.52<t<=-2863.76, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 127 to Candidate_region{branch=126, mut_idx=1, -2863.92<t<=-2861.14, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 128 to Candidate_region{branch=125, mut_idx=3, -2827.16<t<=-2826.32, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 129 to Candidate_region{branch=351, mut_idx=7, -2835.55<t<=-2835.44, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 130 to Candidate_region{branch=123, mut_idx=1, -2928.79<t<=-2858.93, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 131 to Candidate_region{branch=348, mut_idx=0, -3415.31<t<=-3399.63, min_muts=35, W_over_Wmax=0}\n",
            "Attaching tip 132 to Candidate_region{branch=126, mut_idx=0, -2862.96<t<=-2861.14, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 133 to Candidate_region{branch=348, mut_idx=19, -2860.85<t<=-2842, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 134 to Candidate_region{branch=132, mut_idx=0, -2861.99<t<=-2859.77, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 135 to Candidate_region{branch=133, mut_idx=12, -2843.53<t<=-2843, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 136 to Candidate_region{branch=132, mut_idx=2, -2857.65<t<=-2855.71, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 137 to Candidate_region{branch=134, mut_idx=2, -2851.41<t<=-2851, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 138 to Candidate_region{branch=359, mut_idx=19, -2860.85<t<=-2858.62, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 139 to Candidate_region{branch=137, mut_idx=0, -2851.03<t<=-2849, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 140 to Candidate_region{branch=136, mut_idx=0, -2856.98<t<=-2853.4, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 141 to Candidate_region{branch=352, mut_idx=5, -2907.52<t<=-2889, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 142 to Candidate_region{branch=352, mut_idx=0, -2904.03<t<=-2889, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 143 to Candidate_region{branch=353, mut_idx=1, -2863.92<t<=-2863, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 144 to Candidate_region{branch=123, mut_idx=0, -2918.77<t<=-2876, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 145 to Candidate_region{branch=79, mut_idx=1, -13913.9<t<=-13892.4, min_muts=21, W_over_Wmax=0}\n",
            "Attaching tip 146 to Candidate_region{branch=364, mut_idx=0, -3410.73<t<=-3399.63, min_muts=44, W_over_Wmax=0}\n",
            "Attaching tip 147 to Candidate_region{branch=146, mut_idx=24, -3189.96<t<=-3169.2, min_muts=30, W_over_Wmax=0}\n",
            "Attaching tip 148 to Candidate_region{branch=147, mut_idx=33, -2855.16<t<=-2849, min_muts=20, W_over_Wmax=0}\n",
            "Attaching tip 149 to Candidate_region{branch=374, mut_idx=33, -2855.16<t<=-2851.76, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 150 to Candidate_region{branch=146, mut_idx=20, -2875.62<t<=-2870.39, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 151 to Candidate_region{branch=363, mut_idx=0, -2860.85<t<=-2852.46, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 152 to Candidate_region{branch=151, mut_idx=16, -2584.17<t<=-2578, min_muts=38, W_over_Wmax=0}\n",
            "Attaching tip 153 to Candidate_region{branch=131, mut_idx=34, -2857.89<t<=-2838.01, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 154 to Candidate_region{branch=152, mut_idx=17, -2570.78<t<=-2570.09, min_muts=57, W_over_Wmax=0}\n",
            "Attaching tip 155 to Candidate_region{branch=154, mut_idx=2, -2570.33<t<=-2569.85, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 156 to Candidate_region{branch=333, mut_idx=11, -3652.03<t<=-3652.02, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 157 to Candidate_region{branch=117, mut_idx=331, -6574.28<t<=-6574, min_muts=217, W_over_Wmax=0}\n",
            "Attaching tip 158 to Candidate_region{branch=283, mut_idx=0, -9496.02<t<=-9470.02, min_muts=205, W_over_Wmax=0}\n",
            "Attaching tip 159 to Candidate_region{branch=357, mut_idx=0, -3415.31<t<=-3410.73, min_muts=21, W_over_Wmax=0}\n",
            "Attaching tip 160 to Candidate_region{branch=116, mut_idx=1, -4017.41<t<=-4017.41, min_muts=129, W_over_Wmax=0}\n",
            "Attaching tip 161 to Candidate_region{branch=339, mut_idx=23, -4594.65<t<=-4594.21, min_muts=193, W_over_Wmax=0}\n",
            "Attaching tip 162 to Candidate_region{branch=378, mut_idx=6, -2785.69<t<=-2771.9, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 163 to Candidate_region{branch=109, mut_idx=134, -2924.4<t<=-2922, min_muts=93, W_over_Wmax=0}\n",
            "Attaching tip 164 to Candidate_region{branch=161, mut_idx=87, -4170.03<t<=-4166.26, min_muts=170, W_over_Wmax=0}\n",
            "Attaching tip 165 to Candidate_region{branch=157, mut_idx=20, -6258.57<t<=-6245.6, min_muts=305, W_over_Wmax=0}\n",
            "Attaching tip 166 to Candidate_region{branch=160, mut_idx=0, -4017.41<t<=-4017.41, min_muts=121, W_over_Wmax=0}\n",
            "Attaching tip 167 to Candidate_region{branch=25, mut_idx=4, -7308.5<t<=-7308.49, min_muts=141, W_over_Wmax=0}\n",
            "Attaching tip 168 to Candidate_region{branch=331, mut_idx=1, -23396.1<t<=-23392.1, min_muts=334, W_over_Wmax=0}\n",
            "Attaching tip 169 to Candidate_region{branch=165, mut_idx=305, -2193.8<t<=-2191, min_muts=119, W_over_Wmax=0}\n",
            "Attaching tip 170 to Candidate_region{branch=169, mut_idx=26, -2063.56<t<=-2062.05, min_muts=78, W_over_Wmax=0}\n",
            "Attaching tip 171 to Candidate_region{branch=169, mut_idx=93, -1655.79<t<=-1654, min_muts=27, W_over_Wmax=0}\n",
            "Attaching tip 172 to Candidate_region{branch=170, mut_idx=14, -1992.17<t<=-1989.94, min_muts=83, W_over_Wmax=0}\n",
            "Attaching tip 173 to Candidate_region{branch=170, mut_idx=43, -1748.97<t<=-1746.23, min_muts=74, W_over_Wmax=0}\n",
            "Attaching tip 174 to Candidate_region{branch=164, mut_idx=170, -1729.64<t<=-1729, min_muts=159, W_over_Wmax=0}\n",
            "Attaching tip 175 to Candidate_region{branch=174, mut_idx=124, -1461.21<t<=-1461, min_muts=50, W_over_Wmax=0}\n",
            "Attaching tip 176 to Candidate_region{branch=387, mut_idx=6, -4598.27<t<=-4598.1, min_muts=145, W_over_Wmax=0}\n",
            "Attaching tip 177 to Candidate_region{branch=116, mut_idx=173, -4017.01<t<=-4017, min_muts=193, W_over_Wmax=0}\n",
            "Attaching tip 178 to Candidate_region{branch=177, mut_idx=193, -1462.17<t<=-1461, min_muts=241, W_over_Wmax=0}\n",
            "Attaching tip 179 to Candidate_region{branch=404, mut_idx=181, -1540.51<t<=-1534.87, min_muts=92, W_over_Wmax=0}\n",
            "Attaching tip 180 to Candidate_region{branch=395, mut_idx=295, -2352.44<t<=-2324.28, min_muts=82, W_over_Wmax=0}\n",
            "Attaching tip 181 to Candidate_region{branch=401, mut_idx=1, -1727.77<t<=-1727.09, min_muts=77, W_over_Wmax=0}\n",
            "Attaching tip 182 to Candidate_region{branch=339, mut_idx=165, -4566.99<t<=-4566.7, min_muts=364, W_over_Wmax=0}\n",
            "Attaching tip 183 to Candidate_region{branch=176, mut_idx=104, -2668.02<t<=-2660.93, min_muts=142, W_over_Wmax=0}\n",
            "Attaching tip 184 to Candidate_region{branch=178, mut_idx=242, -1461<t<=-1461, min_muts=71, W_over_Wmax=0}\n",
            "Attaching tip 185 to Candidate_region{branch=86, mut_idx=4, -3407.1<t<=-3402, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 186 to Candidate_region{branch=166, mut_idx=1, -4012.47<t<=-3974.58, min_muts=124, W_over_Wmax=0}\n",
            "Attaching tip 187 to Candidate_region{branch=397, mut_idx=34, -1907.64<t<=-1902.98, min_muts=117, W_over_Wmax=0}\n",
            "Attaching tip 188 to Candidate_region{branch=81, mut_idx=16, -4393.3<t<=-4383, min_muts=162, W_over_Wmax=0}\n",
            "Attaching tip 189 to Candidate_region{branch=389, mut_idx=4, -5447.44<t<=-5373.13, min_muts=245, W_over_Wmax=0}\n",
            "Attaching tip 190 to Candidate_region{branch=157, mut_idx=197, -3154.05<t<=-3136, min_muts=208, W_over_Wmax=0}\n",
            "Attaching tip 191 to Candidate_region{branch=188, mut_idx=163, -1574.54<t<=-1567.64, min_muts=307, W_over_Wmax=0}\n",
            "Attaching tip 192 to Candidate_region{branch=389, mut_idx=122, -3150.55<t<=-3140.48, min_muts=263, W_over_Wmax=0}\n",
            "Attaching tip 193 to Candidate_region{branch=396, mut_idx=2, -2190.8<t<=-2179.86, min_muts=139, W_over_Wmax=0}\n",
            "Attaching tip 194 to Candidate_region{branch=184, mut_idx=11, -1368.59<t<=-1365.79, min_muts=213, W_over_Wmax=0}\n",
            "Attaching tip 195 to Candidate_region{branch=145, mut_idx=21, -5847.79<t<=-4968, min_muts=90, W_over_Wmax=0}\n",
            "Attaching tip 196 to Candidate_region{branch=195, mut_idx=90, -1799.48<t<=-1795, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 197 to Candidate_region{branch=196, mut_idx=0, -1798.93<t<=-1798.33, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 198 to Candidate_region{branch=250, mut_idx=58, -9939.43<t<=-9879.3, min_muts=2561, W_over_Wmax=0}\n",
            "Attaching tip 199 to Candidate_region{branch=113, mut_idx=24, -4519.16<t<=-4511.16, min_muts=3376, W_over_Wmax=0}\n",
            "Attaching tip 200 to Candidate_region{branch=302, mut_idx=102, -16172.7<t<=-16071, min_muts=126, W_over_Wmax=0}\n",
            "Attaching tip 201 to Candidate_region{branch=79, mut_idx=4, -13886<t<=-13879, min_muts=21, W_over_Wmax=0}\n",
            "Attaching tip 202 to Candidate_region{branch=342, mut_idx=19, -15199.2<t<=-15162.4, min_muts=319, W_over_Wmax=0}\n",
            "Attaching tip 203 to Candidate_region{branch=167, mut_idx=0, -7308.5<t<=-7290.32, min_muts=78, W_over_Wmax=0}\n",
            "Attaching tip 204 to Candidate_region{branch=343, mut_idx=418, -9554.67<t<=-9496, min_muts=266, W_over_Wmax=0}\n",
            "Attaching tip 205 to Candidate_region{branch=181, mut_idx=47, -1578.83<t<=-1576.9, min_muts=190, W_over_Wmax=0}\n",
            "Attaching tip 206 to Candidate_region{branch=205, mut_idx=187, -276.432<t<=-269.641, min_muts=155, W_over_Wmax=0}\n",
            "Attaching tip 207 to Candidate_region{branch=205, mut_idx=3, -233.632<t<=-225, min_muts=25, W_over_Wmax=0}\n",
            "Attaching tip 208 to Candidate_region{branch=207, mut_idx=13, -223.784<t<=-223.702, min_muts=35, W_over_Wmax=0}\n",
            "Attaching tip 209 to Candidate_region{branch=208, mut_idx=11, -222.769<t<=-222.651, min_muts=99, W_over_Wmax=0}\n",
            "Attaching tip 210 to Candidate_region{branch=182, mut_idx=364, -2191.92<t<=-2191, min_muts=164, W_over_Wmax=0}\n",
            "Attaching tip 211 to Candidate_region{branch=209, mut_idx=97, -220.058<t<=-220.057, min_muts=30, W_over_Wmax=0}\n",
            "Attaching tip 212 to Candidate_region{branch=206, mut_idx=154, -225.404<t<=-225.085, min_muts=122, W_over_Wmax=0}\n",
            "Attaching tip 213 to Candidate_region{branch=211, mut_idx=1, -219.957<t<=-219.014, min_muts=64, W_over_Wmax=0}\n",
            "Attaching tip 214 to Candidate_region{branch=207, mut_idx=10, -223.157<t<=-223.079, min_muts=30, W_over_Wmax=0}\n",
            "Attaching tip 215 to Candidate_region{branch=213, mut_idx=61, -192.841<t<=-191.241, min_muts=132, W_over_Wmax=0}\n",
            "Attaching tip 216 to Candidate_region{branch=212, mut_idx=108, -194.071<t<=-193.8, min_muts=75, W_over_Wmax=0}\n",
            "Attaching tip 217 to Candidate_region{branch=214, mut_idx=3, -220.925<t<=-220.062, min_muts=21, W_over_Wmax=0}\n",
            "Attaching tip 218 to Candidate_region{branch=184, mut_idx=57, -951.183<t<=-950.444, min_muts=109, W_over_Wmax=0}\n",
            "Attaching tip 219 to Candidate_region{branch=218, mut_idx=1, -950.067<t<=-949.217, min_muts=87, W_over_Wmax=0}\n",
            "Attaching tip 220 to Candidate_region{branch=219, mut_idx=10, -814.979<t<=-790.863, min_muts=70, W_over_Wmax=0}\n",
            "Attaching tip 221 to Candidate_region{branch=390, mut_idx=4, -4587.84<t<=-4584.01, min_muts=147, W_over_Wmax=0}\n",
            "Attaching tip 222 to Candidate_region{branch=446, mut_idx=4, -913.109<t<=-898.218, min_muts=67, W_over_Wmax=0}\n",
            "Attaching tip 223 to Candidate_region{branch=175, mut_idx=50, -1461<t<=-1461, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 224 to Candidate_region{branch=216, mut_idx=75, -187.026<t<=-187, min_muts=49, W_over_Wmax=0}\n",
            "Attaching tip 225 to Candidate_region{branch=193, mut_idx=1, -2170.12<t<=-2169.55, min_muts=144, W_over_Wmax=0}\n",
            "Attaching tip 226 to Candidate_region{branch=210, mut_idx=164, -216.804<t<=-216, min_muts=120, W_over_Wmax=0}\n",
            "Built initial tree with 227 tips\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "# Parallelism: 2\n",
            "0.0 M steps/s, Step 0, log_posterior = -414328.45, log_G = -396963.71, log_coal = -19824.73, log_other_priors = 2459.99, num_muts = 24622, T = 1338942.44, t_MRCA = 1852-08-16, n0 = 1000.0, g = 0.0000, mu = 1 * 10^-3 subst / site / year, M_ts = 19109, M_tv = 5513, kappa = 1, pi = [0.31, 0.21, 0.26, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3424, 2307, 2882, 2442], num_muts_ab counts: [827, 4155, 931, 866, 508, 5399, 4163, 540, 429, 966, 5392, 446]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 78.7576 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.0 M steps/s, Step 2270, log_posterior = -385853.06, log_G = -385409.28, log_coal = -2900.64, log_other_priors = 2456.87, num_muts = 24620, T = 1341988.82, t_MRCA = 1852-08-02, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19103, M_tv = 5517, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3423, 2310, 2883, 2439], num_muts_ab counts: [827, 4155, 930, 866, 510, 5398, 4162, 541, 430, 967, 5388, 446]\n",
            "0.0 M steps/s, Step 4540, log_posterior = -386191.79, log_G = -385748.51, log_coal = -2900.14, log_other_priors = 2456.87, num_muts = 24643, T = 1342187.91, t_MRCA = 1852-07-22, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19109, M_tv = 5534, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3424, 2310, 2878, 2443], num_muts_ab counts: [828, 4159, 933, 869, 512, 5398, 4164, 542, 431, 970, 5388, 449]\n",
            "0.0 M steps/s, Step 6810, log_posterior = -386744.73, log_G = -386296.82, log_coal = -2904.77, log_other_priors = 2456.87, num_muts = 24673, T = 1348204.97, t_MRCA = 1852-08-17, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19118, M_tv = 5555, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3426, 2309, 2878, 2442], num_muts_ab counts: [832, 4162, 936, 872, 514, 5397, 4169, 545, 434, 969, 5390, 453]\n",
            "0.0 M steps/s, Step 9080, log_posterior = -386982.31, log_G = -386530.68, log_coal = -2908.49, log_other_priors = 2456.87, num_muts = 24687, T = 1349861.39, t_MRCA = 1852-07-23, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19123, M_tv = 5564, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3427, 2308, 2881, 2439], num_muts_ab counts: [832, 4162, 936, 872, 516, 5400, 4171, 548, 437, 969, 5390, 454]\n",
            "0.0 M steps/s, Step 11350, log_posterior = -387116.50, log_G = -386670.55, log_coal = -2902.82, log_other_priors = 2456.87, num_muts = 24705, T = 1343641.11, t_MRCA = 1852-07-05, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19135, M_tv = 5570, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3425, 2310, 2883, 2437], num_muts_ab counts: [831, 4162, 936, 870, 519, 5406, 4175, 551, 437, 971, 5392, 455]\n",
            "0.0 M steps/s, Step 13620, log_posterior = -387234.44, log_G = -386789.14, log_coal = -2902.17, log_other_priors = 2456.87, num_muts = 24712, T = 1344284.49, t_MRCA = 1852-07-14, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19132, M_tv = 5580, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3426, 2307, 2889, 2433], num_muts_ab counts: [835, 4159, 939, 874, 517, 5403, 4176, 551, 439, 971, 5394, 454]\n",
            "0.0 M steps/s, Step 15890, log_posterior = -387658.20, log_G = -387206.76, log_coal = -2908.31, log_other_priors = 2456.87, num_muts = 24735, T = 1348916.73, t_MRCA = 1852-07-03, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19141, M_tv = 5594, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3423, 2307, 2889, 2436], num_muts_ab counts: [836, 4163, 940, 878, 519, 5399, 4183, 554, 440, 970, 5396, 457]\n",
            "0.0 M steps/s, Step 18160, log_posterior = -386700.86, log_G = -386262.33, log_coal = -2895.39, log_other_priors = 2456.87, num_muts = 24723, T = 1307352.62, t_MRCA = 1852-01-07, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19137, M_tv = 5586, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3423, 2309, 2887, 2436], num_muts_ab counts: [835, 4163, 938, 879, 516, 5402, 4179, 552, 440, 968, 5393, 458]\n",
            "0.0 M steps/s, Step 20430, log_posterior = -386663.80, log_G = -386224.79, log_coal = -2895.87, log_other_priors = 2456.87, num_muts = 24720, T = 1307606.43, t_MRCA = 1851-12-31, n0 = 9528.5, g = -0.0002, mu = 0.64 * 10^-3 subst / site / year, M_ts = 19134, M_tv = 5586, kappa = 3.25, pi = [0.32, 0.21, 0.25, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3421, 2308, 2890, 2436], num_muts_ab counts: [835, 4160, 938, 880, 516, 5403, 4180, 553, 440, 968, 5391, 456]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 193.06 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.0 M steps/s, Step 22700, log_posterior = -399152.56, log_G = -398721.97, log_coal = -2886.86, log_other_priors = 2456.27, num_muts = 25450, T = 1445155.29, t_MRCA = 1687-06-03, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19324, M_tv = 6126, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3466, 2307, 2852, 2430], num_muts_ab counts: [877, 4247, 1008, 950, 585, 5457, 4170, 646, 515, 1028, 5450, 517]\n",
            "0.0 M steps/s, Step 24970, log_posterior = -397945.81, log_G = -397516.95, log_coal = -2885.13, log_other_priors = 2456.27, num_muts = 25378, T = 1437215.81, t_MRCA = 1687-05-27, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19307, M_tv = 6071, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3478, 2267, 2876, 2434], num_muts_ab counts: [915, 4209, 999, 887, 574, 5463, 4188, 628, 509, 1040, 5447, 519]\n",
            "0.0 M steps/s, Step 27240, log_posterior = -398701.90, log_G = -398271.18, log_coal = -2886.99, log_other_priors = 2456.27, num_muts = 25416, T = 1451368.83, t_MRCA = 1668-08-14, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19362, M_tv = 6054, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3497, 2292, 2849, 2417], num_muts_ab counts: [882, 4238, 1013, 894, 575, 5482, 4199, 634, 507, 993, 5443, 556]\n",
            "0.0 M steps/s, Step 29510, log_posterior = -398040.88, log_G = -397620.87, log_coal = -2876.29, log_other_priors = 2456.27, num_muts = 25431, T = 1402760.40, t_MRCA = 1668-08-12, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19379, M_tv = 6052, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3496, 2299, 2845, 2415], num_muts_ab counts: [883, 4234, 1012, 894, 575, 5495, 4203, 632, 504, 991, 5447, 561]\n",
            "0.0 M steps/s, Step 31780, log_posterior = -397986.11, log_G = -397565.95, log_coal = -2876.44, log_other_priors = 2456.27, num_muts = 25425, T = 1404778.29, t_MRCA = 1668-08-05, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19378, M_tv = 6047, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3496, 2299, 2845, 2415], num_muts_ab counts: [883, 4233, 1013, 893, 574, 5493, 4204, 630, 502, 991, 5448, 561]\n",
            "0.0 M steps/s, Step 34050, log_posterior = -397845.26, log_G = -397424.91, log_coal = -2876.62, log_other_priors = 2456.27, num_muts = 25415, T = 1404932.90, t_MRCA = 1668-08-05, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19372, M_tv = 6043, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3496, 2299, 2845, 2415], num_muts_ab counts: [883, 4230, 1010, 891, 575, 5490, 4206, 631, 502, 991, 5446, 560]\n",
            "0.0 M steps/s, Step 36320, log_posterior = -398286.47, log_G = -397866.34, log_coal = -2876.40, log_other_priors = 2456.27, num_muts = 25446, T = 1404949.24, t_MRCA = 1668-08-08, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19392, M_tv = 6054, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3499, 2300, 2845, 2411], num_muts_ab counts: [884, 4236, 1012, 892, 575, 5496, 4209, 632, 505, 991, 5451, 563]\n",
            "0.0 M steps/s, Step 38590, log_posterior = -398257.44, log_G = -397835.65, log_coal = -2878.06, log_other_priors = 2456.27, num_muts = 25446, T = 1402813.71, t_MRCA = 1668-07-31, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19385, M_tv = 6061, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3499, 2291, 2849, 2416], num_muts_ab counts: [885, 4239, 1012, 893, 574, 5488, 4212, 634, 507, 992, 5446, 564]\n",
            "0.0 M steps/s, Step 40860, log_posterior = -398240.97, log_G = -397818.23, log_coal = -2879.01, log_other_priors = 2456.27, num_muts = 25443, T = 1404252.09, t_MRCA = 1668-02-18, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19383, M_tv = 6060, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3499, 2291, 2849, 2416], num_muts_ab counts: [884, 4239, 1013, 892, 574, 5486, 4212, 634, 507, 992, 5446, 564]\n",
            "0.0 M steps/s, Step 43130, log_posterior = -398312.93, log_G = -397890.63, log_coal = -2878.57, log_other_priors = 2456.27, num_muts = 25449, T = 1403965.00, t_MRCA = 1668-03-02, n0 = 12323.7, g = -0.0002, mu = 0.6 * 10^-3 subst / site / year, M_ts = 19392, M_tv = 6057, kappa = 4.3, pi = [0.32, 0.22, 0.24, 0.22], alpha OFF, nu = 1.000, Root Seq counts: [3493, 2297, 2851, 2414], num_muts_ab counts: [884, 4248, 1009, 899, 572, 5484, 4215, 634, 507, 991, 5445, 561]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 259.002 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.0 M steps/s, Step 45400, log_posterior = -397374.23, log_G = -397055.75, log_coal = -2773.20, log_other_priors = 2454.73, num_muts = 25431, T = 1403506.65, t_MRCA = 1668-03-02, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19379, M_tv = 6052, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3498, 2293, 2850, 2414], num_muts_ab counts: [885, 4247, 1007, 898, 572, 5480, 4212, 635, 506, 989, 5440, 560]\n",
            "0.0 M steps/s, Step 47670, log_posterior = -397302.34, log_G = -396985.36, log_coal = -2771.70, log_other_priors = 2454.73, num_muts = 25433, T = 1398208.19, t_MRCA = 1668-03-23, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19383, M_tv = 6050, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3495, 2294, 2851, 2415], num_muts_ab counts: [886, 4248, 1003, 897, 572, 5482, 4212, 635, 508, 989, 5441, 560]\n",
            "0.0 M steps/s, Step 49940, log_posterior = -396979.58, log_G = -396662.76, log_coal = -2771.55, log_other_priors = 2454.73, num_muts = 25413, T = 1397126.57, t_MRCA = 1668-05-10, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19379, M_tv = 6034, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3496, 2289, 2848, 2422], num_muts_ab counts: [884, 4248, 998, 893, 570, 5480, 4209, 633, 507, 990, 5442, 559]\n",
            "0.0 M steps/s, Step 52210, log_posterior = -396922.52, log_G = -396604.88, log_coal = -2772.37, log_other_priors = 2454.73, num_muts = 25408, T = 1397975.67, t_MRCA = 1668-06-01, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19377, M_tv = 6031, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3496, 2289, 2849, 2421], num_muts_ab counts: [884, 4248, 998, 891, 570, 5477, 4210, 631, 508, 990, 5442, 559]\n",
            "0.0 M steps/s, Step 54480, log_posterior = -396753.98, log_G = -396435.53, log_coal = -2773.17, log_other_priors = 2454.73, num_muts = 25395, T = 1398805.51, t_MRCA = 1668-06-12, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19368, M_tv = 6027, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3494, 2289, 2852, 2420], num_muts_ab counts: [883, 4245, 995, 888, 571, 5475, 4210, 631, 510, 991, 5438, 558]\n",
            "0.0 M steps/s, Step 56750, log_posterior = -396630.71, log_G = -396311.99, log_coal = -2773.45, log_other_priors = 2454.73, num_muts = 25386, T = 1399311.21, t_MRCA = 1668-05-17, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19365, M_tv = 6021, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3493, 2290, 2851, 2421], num_muts_ab counts: [881, 4244, 993, 888, 571, 5474, 4210, 631, 509, 990, 5437, 558]\n",
            "0.0 M steps/s, Step 59020, log_posterior = -396608.66, log_G = -396290.34, log_coal = -2773.05, log_other_priors = 2454.73, num_muts = 25384, T = 1399514.45, t_MRCA = 1668-06-26, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19361, M_tv = 6023, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3494, 2291, 2849, 2421], num_muts_ab counts: [880, 4243, 993, 889, 572, 5474, 4210, 632, 509, 990, 5434, 558]\n",
            "0.0 M steps/s, Step 61290, log_posterior = -396451.65, log_G = -396132.38, log_coal = -2773.99, log_other_priors = 2454.73, num_muts = 25375, T = 1397663.45, t_MRCA = 1668-05-28, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19353, M_tv = 6022, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3493, 2293, 2848, 2421], num_muts_ab counts: [879, 4240, 995, 889, 571, 5472, 4208, 631, 508, 991, 5433, 558]\n",
            "0.0 M steps/s, Step 63560, log_posterior = -396319.38, log_G = -396000.01, log_coal = -2774.10, log_other_priors = 2454.73, num_muts = 25365, T = 1398232.59, t_MRCA = 1668-05-28, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19345, M_tv = 6020, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3498, 2289, 2848, 2420], num_muts_ab counts: [879, 4237, 999, 889, 568, 5469, 4207, 631, 506, 990, 5432, 558]\n",
            "0.0 M steps/s, Step 65830, log_posterior = -396238.38, log_G = -395919.13, log_coal = -2773.98, log_other_priors = 2454.73, num_muts = 25359, T = 1398083.49, t_MRCA = 1668-06-12, n0 = 30105.8, g = -0.0001, mu = 0.61 * 10^-3 subst / site / year, M_ts = 19338, M_tv = 6021, kappa = 6.77, pi = [0.31, 0.23, 0.24, 0.23], alpha OFF, nu = 1.000, Root Seq counts: [3497, 2286, 2850, 2422], num_muts_ab counts: [879, 4232, 1000, 887, 569, 5470, 4205, 632, 505, 991, 5431, 558]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zXSO7xfr_fG_"
      },
      "outputs": [],
      "source": [
        "# gget.delphy(aligned_fasta_file, metadata_file)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H8W4ESuFoA9W"
      },
      "source": [
        "___\n",
        "# All done! 🎉\n",
        "\n",
        "### To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.\n",
        "\n",
        "### To further visualize your Delphy output, upload the .dphy file to https://delphy.fathom.info/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kkDLta4bos7Z"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}