{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/delphy_workflows/blob/main/delphy_workflow.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "# Running [Delphy](https://delphy.fathom.info/) is simple as 1. 2. 3.\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/):"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title NCBI Virus filtering options:\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "  if arg == \"True\":\n",
        "      return True\n",
        "  elif arg == \"False\":\n",
        "      return False\n",
        "  elif arg == \"None\" or arg == \"\":\n",
        "      return None\n",
        "  else:\n",
        "      return arg\n",
        "\n",
        "#@markdown ## Virus\n",
        "\n",
        "virus = 'dengue virus type 4'  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "accession = False   #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` argument above is an NCBI accession (starts with 'NC'), e.g. 'NC_045512.2'.\n",
        "\n",
        "#@markdown ## Host\n",
        "\n",
        "host = 'homo sapiens'  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'homo sapiens' (alternative: use the `host_taxid` filter). Input 'None' to disable this filter.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "host_taxid = None  #@param {type:\"raw\"}\n",
        "#@markdown  - NCBI Taxonomy ID of host (e.g., 9443 for primates).\n",
        "host_taxid = arg_str_to_bool(host_taxid)\n",
        "\n",
        "#@markdown ## Sequence completeness\n",
        "\n",
        "annotated = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Indicates whether the sequences should be marked as 'annotated'.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "nuc_completeness = \"None\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Choose between 'partial' or 'complete' nucleotide completeness.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum sequence length, e.g. 6252.\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum sequence length, e.g. 7815.\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require sequences to contain specific proteins (e.g. input 'GPC' or 'L' - include the quotation marks for this filter) or a list of proteins (e.g. input ['GPC', 'L']). Also accepts names of genes or segments.\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to 'True' if the proteins/genes/segments in `has_proteins` should be marked as complete.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum number of 'N' characters allowed in each sequence, e.g. 10.\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "#@markdown ## Gene/peptide/protein counts\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum gene count, e.g. 1.\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum gene count, e.g. 40.\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum peptide count, e.g. 2.\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum peptide count, e.g. 15.\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum protein count, e.g. 2.\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum protein count, e.g. 10.\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "\n",
        "#@markdown ## Geographic location\n",
        "\n",
        "geographic_location = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection, e.g. 'South Africa' or 'Germany'.\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "geographic_region = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic region of sample collection, e.g. 'Africa' or 'Europe'.\n",
        "geographic_region = arg_str_to_bool(geographic_region)\n",
        "\n",
        "#@markdown ## Dates\n",
        "\n",
        "min_collection_date = None  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum collection date, e.g. '2000-01-01'.\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "\n",
        "max_collection_date = None  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum collection date, e.g. '2014-12-04'.\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "\n",
        "min_release_date = None  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum release date of the sequences, e.g. '2000-01-01'.\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "\n",
        "max_release_date = None  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum release date of the sequences, e.g. '2014-12-04'.\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "#@markdown ## Source\n",
        "\n",
        "submitter_country = None  #@param {type:\"string\"}\n",
        "#@markdown  - Country that submitted the sequence, e.g. 'South Africa' or 'Germany'.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to True to return sequences that have been passaged in a laboratory setting.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "source_database = None  #@param {type:\"string\"}\n",
        "#@markdown  - Source database of the sequence, e.g. 'GenBank' or 'RefSeq'.\n",
        "source_database = arg_str_to_bool(source_database)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file(s) here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Select on `Runtime` at the top of this notebook, then click `Run all` and lean back\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "💡 Tip: Click on the folder icon on the left to view/download the files that are being generated.\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title # Generating tree...\n",
        "\n",
        "print(\"1/5 Installing software...\")\n",
        "# Install gget\n",
        "# After the release, this will just be: pip install gget (dependence on biopython will be removed)\n",
        "!pip install -q mysql-connector-python>=8.0.32 biopython\n",
        "!pip install -q git+https://github.com/pachterlab/gget.git@delphy_dev\n",
        "\n",
        "import gget\n",
        "from Bio import SeqIO\n",
        "import pandas as pd\n",
        "import re\n",
        "from datetime import datetime\n",
        "\n",
        "# Delphy threads\n",
        "threads = 2\n",
        "\n",
        "print(\"1/5 Software installation complete.\")\n",
        "\n",
        "# Downloading virus genomes from NCBI Virus\n",
        "print(\"2/5 Download data from NCBI Virus... This might take a minute depending on the internet connection and how busy the NCBI server is.\")\n",
        "gget.ncbi_virus(\n",
        "    virus = virus,\n",
        "    accession = accession,\n",
        "    host = host,\n",
        "    min_seq_length = min_seq_length,\n",
        "    max_seq_length = max_seq_length,\n",
        "    min_gene_count = min_gene_count,\n",
        "    max_gene_count = max_gene_count,\n",
        "    nuc_completeness = nuc_completeness,\n",
        "    has_proteins = has_proteins,\n",
        "    proteins_complete = proteins_complete,\n",
        "    host_taxid = host_taxid,\n",
        "    lab_passaged = lab_passaged,\n",
        "    geographic_region = geographic_region,\n",
        "    geographic_location = geographic_location,\n",
        "    submitter_country = submitter_country,\n",
        "    min_collection_date = min_collection_date,\n",
        "    max_collection_date = max_collection_date,\n",
        "    annotated = annotated,\n",
        "    source_database = source_database,\n",
        "    min_release_date = min_release_date,\n",
        "    max_release_date = max_release_date,\n",
        "    min_mature_peptide_count = min_mature_peptide_count,\n",
        "    max_mature_peptide_count = max_mature_peptide_count,\n",
        "    min_protein_count = min_protein_count,\n",
        "    max_protein_count = max_protein_count,\n",
        "    max_ambiguous_chars = max_ambiguous_chars\n",
        ")\n",
        "print(\"2/5 Data download from NCBI virus complete.\")\n",
        "\n",
        "# Merging sequencing and metadata files if additional file(s) were provided\n",
        "ncbi_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences.fasta\"\n",
        "ncbi_metadata = f\"{'_'.join(str(virus).split(' '))}_metadata.csv\"\n",
        "\n",
        "if fasta_file:\n",
        "  print(\"Adding user-provided fasta file and metadata to the data from NCBI Virus...\")\n",
        "\n",
        "  # Combine sequence files\n",
        "  combined_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences_combined.fasta\"\n",
        "  !cat $ncbi_fasta_file $fasta_file > $combined_fasta_file\n",
        "  input_fasta_file = combined_fasta_file\n",
        "\n",
        "  # Combine metadata\n",
        "  combined_metadata_file = f\"{'_'.join(virus.split(' '))}_metadata_combined.csv\"\n",
        "  ncbi_metadata_df = pd.read_csv(ncbi_metadata)\n",
        "  if metadata_csv:\n",
        "    # Combine provided metadata and NCBI metadata csv files\n",
        "    user_metada_df = pd.read_csv(metadata_csv)\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  else:\n",
        "    # Extract sequence accessions from the provided FASTA file\n",
        "    headers = [record.id.split(\" \")[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "\n",
        "    # Create a metadata dataframe with the accessions from the FASTA file and the provided metadata\n",
        "    user_metada_df = pd.DataFrame(headers, columns=[\"Accession\"])\n",
        "    for key, value in metadata.items():\n",
        "      user_metada_df[key] = value\n",
        "\n",
        "    # Combine with NCBI metadata\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  print(\"Merging user-provided and NCBI Virus data complete.\")\n",
        "\n",
        "else:\n",
        "  input_fasta_file = ncbi_fasta_file\n",
        "  metadata_file = ncbi_metadata\n",
        "\n",
        "# Create MSA\n",
        "print(\"3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...\")\n",
        "\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "\n",
        "# # Option 1: Using the MUSCLE algorithm (this works well for a few hundred sequences, but is too slow when dealing with a few thousand sequences)\n",
        "# gget.muscle(input_fasta_file, super5=True, out=aligned_fasta_file)\n",
        "\n",
        "# Option 2: Using mafft\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.mafft(input_fasta_file, out=aligned_fasta_file)`\n",
        "\n",
        "#Install MAFFT\n",
        "!apt-get install -qq -y mafft\n",
        "\n",
        "# Aligning sequences to each other using mafft\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "!mafft \\\n",
        "  --quiet \\\n",
        "  --auto \\\n",
        "  --thread 2 \\\n",
        "  $input_fasta_file > $aligned_fasta_file\n",
        "\n",
        "print(\"3/5 MSA complete.\")\n",
        "\n",
        "\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.delphy(aligned_fasta_file, metadata_file)`\n",
        "\n",
        "# Adjust the headers in the aligned fasta file to match header format required by Delphy (accession|YYYY-MM-DD):\n",
        "print(\"4/5 Reformatting sequence files to match Delphy format...\")\n",
        "\n",
        "# Reformat collection date\n",
        "default_day = '01'\n",
        "default_month = '01'\n",
        "def extract_and_format_date(date_string):\n",
        "    # Define regular expressions for various date formats\n",
        "    year_only = re.compile(r'(?P<year>\\d{4})')\n",
        "    year_month = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})')\n",
        "    full_date = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})[-/.](?P<day>\\d{1,2})')\n",
        "\n",
        "    # Try to match the full date first\n",
        "    match = full_date.search(date_string)\n",
        "    if match:\n",
        "        year = match.group('year')\n",
        "        month = match.group('month').zfill(2)\n",
        "        day = match.group('day').zfill(2)\n",
        "    else:\n",
        "        # Try to match year and month\n",
        "        match = year_month.search(date_string)\n",
        "        if match:\n",
        "            year = match.group('year')\n",
        "            month = match.group('month').zfill(2)\n",
        "            day = default_day\n",
        "        else:\n",
        "            # Try to match only the year\n",
        "            match = year_only.search(date_string)\n",
        "            if match:\n",
        "                year = match.group('year')\n",
        "                month = default_month\n",
        "                day = default_day\n",
        "            else:\n",
        "                # If no match, return None\n",
        "                return None\n",
        "\n",
        "    # Format the extracted date into YYYY-MM-DD\n",
        "    formatted_date = f\"{year}-{month}-{day}\"\n",
        "\n",
        "    try:\n",
        "        # Validate date by trying to convert it to a datetime object\n",
        "        datetime.strptime(formatted_date, '%Y-%m-%d')\n",
        "    except ValueError:\n",
        "        return None  # Return None if the date is invalid\n",
        "\n",
        "    return formatted_date\n",
        "\n",
        "def update_fasta_headers(fasta_file, csv_file, output_fasta):\n",
        "    # Read the CSV file into a pandas DataFrame\n",
        "    df = pd.read_csv(csv_file)\n",
        "\n",
        "    # Create a dictionary from the DataFrame to map accession to date\n",
        "    accession_to_date = pd.Series(df['Collection Date'].values, index=df['accession']).to_dict()\n",
        "\n",
        "    # Open the input fasta file and output fasta file\n",
        "    with open(fasta_file) as fasta_input, open(output_fasta, 'w') as fasta_output:\n",
        "        # Iterate through each sequence record in the fasta file\n",
        "        for record in SeqIO.parse(fasta_input, 'fasta'):\n",
        "            accession = record.id\n",
        "\n",
        "            # Check if the accession is in the pandas dictionary and has a non-NaN date\n",
        "            if accession in accession_to_date and pd.notna(accession_to_date[accession]):\n",
        "                date = accession_to_date[accession]\n",
        "\n",
        "                # Format the date if necessary\n",
        "                formatted_date = extract_and_format_date(date)\n",
        "\n",
        "                if formatted_date is None:\n",
        "                  # Skip the entry if date is NaN or accession not found\n",
        "                  print(f\"Skipping accession {accession} due to unrecognized date format: '{date}'\")\n",
        "                  continue\n",
        "\n",
        "                # Update the seq header\n",
        "                record.id = f\"{accession}|{formatted_date}\"\n",
        "                record.description = ''  # Remove the original description to avoid duplication\n",
        "            else:\n",
        "                # Skip the entry if date is NaN or accession not found\n",
        "                print(f\"Skipping accession {accession} due to missing or NaN date.\")\n",
        "                continue\n",
        "\n",
        "            # Write the updated record to the output fasta file\n",
        "            SeqIO.write(record, fasta_output, 'fasta')\n",
        "\n",
        "aligned_fasta_file_clean = f\"{'_'.join(str(virus).split(' '))}_aligned_headers_adjusted.afa\"\n",
        "update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean)\n",
        "\n",
        "print(\"4/5 Reformatting complete.\")\n",
        "\n",
        "# Run Delphy\n",
        "print(\"5/5 Running Delphy...\")\n",
        "\n",
        "# Download delphy binary\n",
        "!wget https://github.com/broadinstitute/delphy/releases/download/0.9995/delphy-ubuntu-x86_64\n",
        "\n",
        "# Give permissions\n",
        "!chmod u+x ./delphy-ubuntu-x86_64\n",
        "\n",
        "beast_log_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_log.txt\"\n",
        "delphy_beast_tree_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_tree.nwk\"\n",
        "dphy_out = f\"{'_'.join(str(virus).split(' '))}_delphy_out.dphy\"\n",
        "\n",
        "!./delphy-ubuntu-x86_64 \\\n",
        "  --v0-threads $threads \\\n",
        "  --v0-in-fasta $aligned_fasta_file_clean \\\n",
        "  --v0-out-log-file $beast_log_out \\\n",
        "  --v0-out-trees-file $delphy_beast_tree_out \\\n",
        "  --v0-out-delphy-file $dphy_out\n",
        "\n",
        "\n",
        "# Display a message when done\n",
        "from IPython.display import HTML\n",
        "\n",
        "def done_message():\n",
        "    display(HTML(\"\"\"\n",
        "    <h1>All done! 🎉</h1>\n",
        "    <h3>To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.</h3>\n",
        "    <h3>To further visualize your Delphy output, upload the <code>.dphy</code> file to <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a></h3>\n",
        "    \"\"\"))\n",
        "\n",
        "done_message()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "HuVr7cb0x5VF",
        "outputId": "466d9b4b-9fd5-451c-c039-5450045d0ab2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Installing software...\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for gget (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Software installation complete.\n",
            "Download data from NCBI Virus... This might take a minute depending on the internet connection and how busy the NCBI server is.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "New version of client (16.32.0) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets.\n",
            "INFO:gget.utils:3450 sequences passed the provided filters.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Data download from NCBI virus is complete.\n",
            "Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package fonts-lato.\n",
            "(Reading database ... 123629 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-lato_2.0-2.1_all.deb ...\n",
            "Unpacking fonts-lato (2.0-2.1) ...\n",
            "Selecting previously unselected package netbase.\n",
            "Preparing to unpack .../01-netbase_6.3_all.deb ...\n",
            "Unpacking netbase (6.3) ...\n",
            "Selecting previously unselected package libclone-perl.\n",
            "Preparing to unpack .../02-libclone-perl_0.45-1build3_amd64.deb ...\n",
            "Unpacking libclone-perl (0.45-1build3) ...\n",
            "Selecting previously unselected package libdata-dump-perl.\n",
            "Preparing to unpack .../03-libdata-dump-perl_1.25-1_all.deb ...\n",
            "Unpacking libdata-dump-perl (1.25-1) ...\n",
            "Selecting previously unselected package libencode-locale-perl.\n",
            "Preparing to unpack .../04-libencode-locale-perl_1.05-1.1_all.deb ...\n",
            "Unpacking libencode-locale-perl (1.05-1.1) ...\n",
            "Selecting previously unselected package libhttp-date-perl.\n",
            "Preparing to unpack .../05-libhttp-date-perl_6.05-1_all.deb ...\n",
            "Unpacking libhttp-date-perl (6.05-1) ...\n",
            "Selecting previously unselected package libfile-listing-perl.\n",
            "Preparing to unpack .../06-libfile-listing-perl_6.14-1_all.deb ...\n",
            "Unpacking libfile-listing-perl (6.14-1) ...\n",
            "Selecting previously unselected package libfont-afm-perl.\n",
            "Preparing to unpack .../07-libfont-afm-perl_1.20-3_all.deb ...\n",
            "Unpacking libfont-afm-perl (1.20-3) ...\n",
            "Selecting previously unselected package libhtml-tagset-perl.\n",
            "Preparing to unpack .../08-libhtml-tagset-perl_3.20-4_all.deb ...\n",
            "Unpacking libhtml-tagset-perl (3.20-4) ...\n",
            "Selecting previously unselected package liburi-perl.\n",
            "Preparing to unpack .../09-liburi-perl_5.10-1_all.deb ...\n",
            "Unpacking liburi-perl (5.10-1) ...\n",
            "Selecting previously unselected package libhtml-parser-perl:amd64.\n",
            "Preparing to unpack .../10-libhtml-parser-perl_3.76-1build2_amd64.deb ...\n",
            "Unpacking libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Selecting previously unselected package libio-html-perl.\n",
            "Preparing to unpack .../11-libio-html-perl_1.004-2_all.deb ...\n",
            "Unpacking libio-html-perl (1.004-2) ...\n",
            "Selecting previously unselected package liblwp-mediatypes-perl.\n",
            "Preparing to unpack .../12-liblwp-mediatypes-perl_6.04-1_all.deb ...\n",
            "Unpacking liblwp-mediatypes-perl (6.04-1) ...\n",
            "Selecting previously unselected package libhttp-message-perl.\n",
            "Preparing to unpack .../13-libhttp-message-perl_6.36-1_all.deb ...\n",
            "Unpacking libhttp-message-perl (6.36-1) ...\n",
            "Selecting previously unselected package libhtml-form-perl.\n",
            "Preparing to unpack .../14-libhtml-form-perl_6.07-1_all.deb ...\n",
            "Unpacking libhtml-form-perl (6.07-1) ...\n",
            "Selecting previously unselected package libhtml-tree-perl.\n",
            "Preparing to unpack .../15-libhtml-tree-perl_5.07-2_all.deb ...\n",
            "Unpacking libhtml-tree-perl (5.07-2) ...\n",
            "Selecting previously unselected package libhtml-format-perl.\n",
            "Preparing to unpack .../16-libhtml-format-perl_2.12-1.1_all.deb ...\n",
            "Unpacking libhtml-format-perl (2.12-1.1) ...\n",
            "Selecting previously unselected package libhttp-cookies-perl.\n",
            "Preparing to unpack .../17-libhttp-cookies-perl_6.10-1_all.deb ...\n",
            "Unpacking libhttp-cookies-perl (6.10-1) ...\n",
            "Selecting previously unselected package libhttp-daemon-perl.\n",
            "Preparing to unpack .../18-libhttp-daemon-perl_6.13-1ubuntu0.1_all.deb ...\n",
            "Unpacking libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libhttp-negotiate-perl.\n",
            "Preparing to unpack .../19-libhttp-negotiate-perl_6.01-1_all.deb ...\n",
            "Unpacking libhttp-negotiate-perl (6.01-1) ...\n",
            "Selecting previously unselected package perl-openssl-defaults:amd64.\n",
            "Preparing to unpack .../20-perl-openssl-defaults_5build2_amd64.deb ...\n",
            "Unpacking perl-openssl-defaults:amd64 (5build2) ...\n",
            "Selecting previously unselected package libnet-ssleay-perl:amd64.\n",
            "Preparing to unpack .../21-libnet-ssleay-perl_1.92-1build2_amd64.deb ...\n",
            "Unpacking libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Selecting previously unselected package libio-socket-ssl-perl.\n",
            "Preparing to unpack .../22-libio-socket-ssl-perl_2.074-2_all.deb ...\n",
            "Unpacking libio-socket-ssl-perl (2.074-2) ...\n",
            "Selecting previously unselected package libnet-http-perl.\n",
            "Preparing to unpack .../23-libnet-http-perl_6.22-1_all.deb ...\n",
            "Unpacking libnet-http-perl (6.22-1) ...\n",
            "Selecting previously unselected package libtry-tiny-perl.\n",
            "Preparing to unpack .../24-libtry-tiny-perl_0.31-1_all.deb ...\n",
            "Unpacking libtry-tiny-perl (0.31-1) ...\n",
            "Selecting previously unselected package libwww-robotrules-perl.\n",
            "Preparing to unpack .../25-libwww-robotrules-perl_6.02-1_all.deb ...\n",
            "Unpacking libwww-robotrules-perl (6.02-1) ...\n",
            "Selecting previously unselected package libwww-perl.\n",
            "Preparing to unpack .../26-libwww-perl_6.61-1_all.deb ...\n",
            "Unpacking libwww-perl (6.61-1) ...\n",
            "Selecting previously unselected package liblwp-protocol-https-perl.\n",
            "Preparing to unpack .../27-liblwp-protocol-https-perl_6.10-1_all.deb ...\n",
            "Unpacking liblwp-protocol-https-perl (6.10-1) ...\n",
            "Selecting previously unselected package libnet-smtp-ssl-perl.\n",
            "Preparing to unpack .../28-libnet-smtp-ssl-perl_1.04-1_all.deb ...\n",
            "Unpacking libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Selecting previously unselected package libmailtools-perl.\n",
            "Preparing to unpack .../29-libmailtools-perl_2.21-1_all.deb ...\n",
            "Unpacking libmailtools-perl (2.21-1) ...\n",
            "Selecting previously unselected package rubygems-integration.\n",
            "Preparing to unpack .../30-rubygems-integration_1.18_all.deb ...\n",
            "Unpacking rubygems-integration (1.18) ...\n",
            "Selecting previously unselected package ruby3.0.\n",
            "Preparing to unpack .../31-ruby3.0_3.0.2-7ubuntu2.7_amd64.deb ...\n",
            "Unpacking ruby3.0 (3.0.2-7ubuntu2.7) ...\n",
            "Selecting previously unselected package ruby-rubygems.\n",
            "Preparing to unpack .../32-ruby-rubygems_3.3.5-2_all.deb ...\n",
            "Unpacking ruby-rubygems (3.3.5-2) ...\n",
            "Selecting previously unselected package ruby.\n",
            "Preparing to unpack .../33-ruby_1%3a3.0~exp1_amd64.deb ...\n",
            "Unpacking ruby (1:3.0~exp1) ...\n",
            "Selecting previously unselected package rake.\n",
            "Preparing to unpack .../34-rake_13.0.6-2_all.deb ...\n",
            "Unpacking rake (13.0.6-2) ...\n",
            "Selecting previously unselected package ruby-net-telnet.\n",
            "Preparing to unpack .../35-ruby-net-telnet_0.1.1-2_all.deb ...\n",
            "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
            "Selecting previously unselected package ruby-webrick.\n",
            "Preparing to unpack .../36-ruby-webrick_1.7.0-3ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Selecting previously unselected package ruby-xmlrpc.\n",
            "Preparing to unpack .../37-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libruby3.0:amd64.\n",
            "Preparing to unpack .../38-libruby3.0_3.0.2-7ubuntu2.7_amd64.deb ...\n",
            "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.7) ...\n",
            "Selecting previously unselected package lynx-common.\n",
            "Preparing to unpack .../39-lynx-common_2.9.0dev.10-1_all.deb ...\n",
            "Unpacking lynx-common (2.9.0dev.10-1) ...\n",
            "Selecting previously unselected package mafft.\n",
            "Preparing to unpack .../40-mafft_7.490-1_amd64.deb ...\n",
            "Unpacking mafft (7.490-1) ...\n",
            "Selecting previously unselected package mailcap.\n",
            "Preparing to unpack .../41-mailcap_3.70+nmu1ubuntu1_all.deb ...\n",
            "Unpacking mailcap (3.70+nmu1ubuntu1) ...\n",
            "Selecting previously unselected package mime-support.\n",
            "Preparing to unpack .../42-mime-support_3.66_all.deb ...\n",
            "Unpacking mime-support (3.66) ...\n",
            "Selecting previously unselected package libauthen-sasl-perl.\n",
            "Preparing to unpack .../43-libauthen-sasl-perl_2.1600-1.1_all.deb ...\n",
            "Unpacking libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Selecting previously unselected package lynx.\n",
            "Preparing to unpack .../44-lynx_2.9.0dev.10-1_amd64.deb ...\n",
            "Unpacking lynx (2.9.0dev.10-1) ...\n",
            "Setting up libhttp-date-perl (6.05-1) ...\n",
            "Setting up fonts-lato (2.0-2.1) ...\n",
            "Setting up libfile-listing-perl (6.14-1) ...\n",
            "Setting up libfont-afm-perl (1.20-3) ...\n",
            "Setting up mafft (7.490-1) ...\n",
            "Setting up libclone-perl (0.45-1build3) ...\n",
            "Setting up libhtml-tagset-perl (3.20-4) ...\n",
            "Setting up libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Setting up liblwp-mediatypes-perl (6.04-1) ...\n",
            "Setting up libtry-tiny-perl (0.31-1) ...\n",
            "Setting up perl-openssl-defaults:amd64 (5build2) ...\n",
            "Setting up libencode-locale-perl (1.05-1.1) ...\n",
            "Setting up rubygems-integration (1.18) ...\n",
            "Setting up libdata-dump-perl (1.25-1) ...\n",
            "Setting up ruby-net-telnet (0.1.1-2) ...\n",
            "Setting up libio-html-perl (1.004-2) ...\n",
            "Setting up lynx-common (2.9.0dev.10-1) ...\n",
            "Setting up ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Setting up netbase (6.3) ...\n",
            "Setting up lynx (2.9.0dev.10-1) ...\n",
            "update-alternatives: using /usr/bin/lynx to provide /usr/bin/www-browser (www-browser) in auto mode\n",
            "Setting up mailcap (3.70+nmu1ubuntu1) ...\n",
            "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Setting up liburi-perl (5.10-1) ...\n",
            "Setting up libhttp-message-perl (6.36-1) ...\n",
            "Setting up libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Setting up libhttp-negotiate-perl (6.01-1) ...\n",
            "Setting up mime-support (3.66) ...\n",
            "Setting up libhttp-cookies-perl (6.10-1) ...\n",
            "Setting up libnet-http-perl (6.22-1) ...\n",
            "Setting up libwww-robotrules-perl (6.02-1) ...\n",
            "Setting up libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Setting up libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Setting up libio-socket-ssl-perl (2.074-2) ...\n",
            "Setting up libhtml-form-perl (6.07-1) ...\n",
            "Setting up libhtml-tree-perl (5.07-2) ...\n",
            "Setting up libhtml-format-perl (2.12-1.1) ...\n",
            "Setting up libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Setting up libmailtools-perl (2.21-1) ...\n",
            "Setting up libwww-perl (6.61-1) ...\n",
            "Setting up ruby3.0 (3.0.2-7ubuntu2.7) ...\n",
            "Setting up liblwp-protocol-https-perl (6.10-1) ...\n",
            "Setting up ruby (1:3.0~exp1) ...\n",
            "Setting up rake (13.0.6-2) ...\n",
            "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.7) ...\n",
            "Setting up ruby-rubygems (3.3.5-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kkDLta4bos7Z"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}