{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/delphy_workflows/blob/main/delphy_workflow_sars_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "# Running [Delphy](https://github.com/broadinstitute/delphy/blob/main/delphywp.pdf) is simple as 1. 2. 3.\n",
        "Generate a phylogenetic tree using [Delphy](https://github.com/broadinstitute/delphy/blob/main/delphywp.pdf) on sequences downloaded from the [NCBI Virus database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/) (optional: upload your own sequences which will be added to the analysis).\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title NCBI Virus filtering options:\n",
        "#@markdown #### **NOTE: Set any filter to `None` to disable it.**\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "  if arg == \"True\":\n",
        "      return True\n",
        "  elif arg == \"False\":\n",
        "      return False\n",
        "  elif arg == \"None\" or arg == \"\":\n",
        "      return None\n",
        "  else:\n",
        "      return arg\n",
        "\n",
        "#@markdown ## **Virus**\n",
        "\n",
        "virus = 'sars-cov-2'  #@param {type:\"string\"}\n",
        "#@markdown  - Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "accession = False   #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` argument above is an NCBI accession (starts with 'NC'), e.g. 'NC_045512.2'.\n",
        "\n",
        "#@markdown ## **Host**\n",
        "\n",
        "host = 'homo sapiens'  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'homo sapiens' (alternative: use the `host_taxid` filter below). Input `None` to disable filtering by host.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "host_taxid = None  #@param {type:\"raw\"}\n",
        "#@markdown  - NCBI Taxonomy ID of host (e.g., 9443 for primates).\n",
        "host_taxid = arg_str_to_bool(host_taxid)\n",
        "\n",
        "#@markdown ## **Sequence completeness**\n",
        "\n",
        "annotated = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences marked as 'annotated'; set to `False` to only return sequences NOT marked as 'annotated'.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "nuc_completeness = \"complete\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Choose between 'partial' or 'complete' nucleotide completeness.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum sequence length, e.g. 6252.\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum sequence length, e.g. 7815.\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require sequences to contain specific proteins (e.g. 'GPC' **- include the quotation marks**) or a list of proteins (e.g. ['GPC', 'L']). Also accepts names of genes or segments.\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = False    #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if the proteins/genes/segments in `has_proteins` should be marked as 'complete'.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum number of 'N' characters allowed in each sequence, e.g. 10.\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "#@markdown ## **Gene/peptide/protein counts**\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum gene count, e.g. 1.\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum gene count, e.g. 40.\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum peptide count, e.g. 2.\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum peptide count, e.g. 15.\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum protein count, e.g. 2.\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum protein count, e.g. 10.\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "\n",
        "#@markdown ## **Geographic location**\n",
        "\n",
        "geographic_location = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection, e.g. 'South Africa' or 'Germany'.\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "geographic_region = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic region of sample collection, e.g. 'Africa' or 'Europe'.\n",
        "geographic_region = arg_str_to_bool(geographic_region)\n",
        "\n",
        "#@markdown ## **Dates**\n",
        "#@markdown All dates should be supplied in YYYY-MM-DD format.\n",
        "min_collection_date = \"2025-01-15\"  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum collection date, e.g. '2000-01-01'.\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "\n",
        "max_collection_date = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum collection date, e.g. '2014-12-04'.\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "\n",
        "min_release_date = \"2025-01-01\"  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum release date of the sequences, e.g. '2000-01-01'.\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "\n",
        "max_release_date = None  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum release date of the sequences, e.g. '2014-12-04'.\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "#@markdown ## **Source**\n",
        "\n",
        "submitter_country = None  #@param {type:\"string\"}\n",
        "#@markdown  - Country that submitted the sequence, e.g. 'South Africa' or 'Germany'.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences that have been passaged in a laboratory setting; set to `False` to only return sequences that have NOT been passaged in a laboratory setting.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "source_database = None  #@param {type:\"string\"}\n",
        "#@markdown  - Source database of the sequence, e.g. 'GenBank' or 'RefSeq'.\n",
        "source_database = arg_str_to_bool(source_database)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file(s) here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Select on `Runtime` at the top of this notebook, then click `Run all` and lean back\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "💡 Tip: Click on the folder icon on the left to view/download the files that are being generated.\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "#@title # Generating tree...\n",
        "from IPython.display import HTML\n",
        "def log_message(text):\n",
        "    display(HTML(f\"<h2 style='color: green;'>{text}</h2>\"))\n",
        "\n",
        "log_message(\"1/5 Installing software...\")\n",
        "# Install gget\n",
        "# After the release, this will just be: pip install gget (dependence on biopython will be removed)\n",
        "!pip install -q mysql-connector-python biopython\n",
        "!pip install -q git+https://github.com/pachterlab/gget.git@delphy_dev\n",
        "\n",
        "import gget\n",
        "from Bio import SeqIO\n",
        "import pandas as pd\n",
        "import re\n",
        "from datetime import datetime\n",
        "\n",
        "log_message(\"1/5 Software installation complete.\")\n",
        "\n",
        "# Downloading virus genomes from NCBI Virus\n",
        "log_message(\"2/5 Download data from NCBI Virus... This might take a few minutes depending on the internet connection and how busy the NCBI server is.\")\n",
        "gget.ncbi_virus(\n",
        "    virus = virus,\n",
        "    accession = accession,\n",
        "    host = host,\n",
        "    min_seq_length = min_seq_length,\n",
        "    max_seq_length = max_seq_length,\n",
        "    min_gene_count = min_gene_count,\n",
        "    max_gene_count = max_gene_count,\n",
        "    nuc_completeness = nuc_completeness,\n",
        "    has_proteins = has_proteins,\n",
        "    proteins_complete = proteins_complete,\n",
        "    host_taxid = host_taxid,\n",
        "    lab_passaged = lab_passaged,\n",
        "    geographic_region = geographic_region,\n",
        "    geographic_location = geographic_location,\n",
        "    submitter_country = submitter_country,\n",
        "    min_collection_date = min_collection_date,\n",
        "    max_collection_date = max_collection_date,\n",
        "    annotated = annotated,\n",
        "    source_database = source_database,\n",
        "    min_release_date = min_release_date,\n",
        "    max_release_date = max_release_date,\n",
        "    min_mature_peptide_count = min_mature_peptide_count,\n",
        "    max_mature_peptide_count = max_mature_peptide_count,\n",
        "    min_protein_count = min_protein_count,\n",
        "    max_protein_count = max_protein_count,\n",
        "    max_ambiguous_chars = max_ambiguous_chars\n",
        ")\n",
        "log_message(\"2/5 Data download from NCBI Virus complete.\")\n",
        "\n",
        "# Merging sequencing and metadata files if additional file(s) were provided\n",
        "ncbi_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences.fasta\"\n",
        "ncbi_metadata = f\"{'_'.join(str(virus).split(' '))}_metadata.csv\"\n",
        "\n",
        "if fasta_file:\n",
        "  log_message(\"Adding user-provided fasta file and metadata to the data from NCBI Virus...\")\n",
        "\n",
        "  # Combine sequence files\n",
        "  combined_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences_combined.fasta\"\n",
        "  !cat $ncbi_fasta_file $fasta_file > $combined_fasta_file\n",
        "  input_fasta_file = combined_fasta_file\n",
        "\n",
        "  # Combine metadata\n",
        "  combined_metadata_file = f\"{'_'.join(virus.split(' '))}_metadata_combined.csv\"\n",
        "  ncbi_metadata_df = pd.read_csv(ncbi_metadata)\n",
        "  if metadata_csv:\n",
        "    # Combine provided metadata and NCBI metadata csv files\n",
        "    user_metada_df = pd.read_csv(metadata_csv)\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  else:\n",
        "    # Extract sequence accessions from the provided FASTA file\n",
        "    headers = [record.id.split(\" \")[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "\n",
        "    # Create a metadata dataframe with the accessions from the FASTA file and the provided metadata\n",
        "    user_metada_df = pd.DataFrame(headers, columns=[\"Accession\"])\n",
        "    for key, value in metadata.items():\n",
        "      user_metada_df[key] = value\n",
        "\n",
        "    # Combine with NCBI metadata\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  log_message(\"Merging user-provided and NCBI Virus data complete.\")\n",
        "\n",
        "else:\n",
        "  input_fasta_file = ncbi_fasta_file\n",
        "  metadata_file = ncbi_metadata\n",
        "\n",
        "# Create MSA\n",
        "log_message(\"3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...\")\n",
        "\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "\n",
        "# # Option 1: Using the MUSCLE algorithm (this works well for a few hundred sequences, but is too slow when dealing with a few thousand sequences)\n",
        "# gget.muscle(input_fasta_file, super5=True, out=aligned_fasta_file)\n",
        "\n",
        "# Option 2: Using mafft\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.mafft(input_fasta_file, out=aligned_fasta_file)`\n",
        "\n",
        "#Install MAFFT\n",
        "!apt-get install -qq -y mafft\n",
        "\n",
        "# Aligning sequences to each other using mafft\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "!mafft \\\n",
        "  --quiet \\\n",
        "  --auto \\\n",
        "  --thread 2 \\\n",
        "  $input_fasta_file > $aligned_fasta_file\n",
        "\n",
        "log_message(\"3/5 MSA complete.\")\n",
        "\n",
        "\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.delphy(aligned_fasta_file, metadata_file)`\n",
        "\n",
        "# Adjust the headers in the aligned fasta file to match header format required by Delphy (accession|YYYY-MM-DD):\n",
        "log_message(\"4/5 Reformatting sequence files to match Delphy format...\")\n",
        "\n",
        "# Reformat collection date\n",
        "default_day = '01'\n",
        "default_month = '01'\n",
        "def extract_and_format_date(date_string):\n",
        "    # Define regular expressions for various date formats\n",
        "    year_only = re.compile(r'(?P<year>\\d{4})')\n",
        "    year_month = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})')\n",
        "    full_date = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})[-/.](?P<day>\\d{1,2})')\n",
        "\n",
        "    # Try to match the full date first\n",
        "    match = full_date.search(date_string)\n",
        "    if match:\n",
        "        year = match.group('year')\n",
        "        month = match.group('month').zfill(2)\n",
        "        day = match.group('day').zfill(2)\n",
        "    else:\n",
        "        # Try to match year and month\n",
        "        match = year_month.search(date_string)\n",
        "        if match:\n",
        "            year = match.group('year')\n",
        "            month = match.group('month').zfill(2)\n",
        "            day = default_day\n",
        "        else:\n",
        "            # Try to match only the year\n",
        "            match = year_only.search(date_string)\n",
        "            if match:\n",
        "                year = match.group('year')\n",
        "                month = default_month\n",
        "                day = default_day\n",
        "            else:\n",
        "                # If no match, return None\n",
        "                return None\n",
        "\n",
        "    # Format the extracted date into YYYY-MM-DD\n",
        "    formatted_date = f\"{year}-{month}-{day}\"\n",
        "\n",
        "    try:\n",
        "        # Validate date by trying to convert it to a datetime object\n",
        "        datetime.strptime(formatted_date, '%Y-%m-%d')\n",
        "    except ValueError:\n",
        "        return None  # Return None if the date is invalid\n",
        "\n",
        "    return formatted_date\n",
        "\n",
        "def update_fasta_headers(fasta_file, csv_file, output_fasta):\n",
        "    # Read the CSV file into a pandas DataFrame\n",
        "    df = pd.read_csv(csv_file)\n",
        "\n",
        "    # Create a dictionary from the DataFrame to map accession to date\n",
        "    accession_to_date = pd.Series(df['Collection Date'].values, index=df['accession']).to_dict()\n",
        "\n",
        "    # Open the input fasta file and output fasta file\n",
        "    with open(fasta_file) as fasta_input, open(output_fasta, 'w') as fasta_output:\n",
        "        # Iterate through each sequence record in the fasta file\n",
        "        for record in SeqIO.parse(fasta_input, 'fasta'):\n",
        "            accession = record.id\n",
        "\n",
        "            # Check if the accession is in the pandas dictionary and has a non-NaN date\n",
        "            if accession in accession_to_date and pd.notna(accession_to_date[accession]):\n",
        "                date = accession_to_date[accession]\n",
        "\n",
        "                # Format the date if necessary\n",
        "                formatted_date = extract_and_format_date(date)\n",
        "\n",
        "                if formatted_date is None:\n",
        "                  # Skip the entry if date is NaN or accession not found\n",
        "                  print(f\"Skipping accession {accession} due to unrecognized date format: '{date}'\")\n",
        "                  continue\n",
        "\n",
        "                # Update the seq header\n",
        "                record.id = f\"{accession}|{formatted_date}\"\n",
        "                record.description = ''  # Remove the original description to avoid duplication\n",
        "            else:\n",
        "                # Skip the entry if date is NaN or accession not found\n",
        "                print(f\"Skipping accession {accession} due to missing or NaN date.\")\n",
        "                continue\n",
        "\n",
        "            # Write the updated record to the output fasta file\n",
        "            SeqIO.write(record, fasta_output, 'fasta')\n",
        "\n",
        "aligned_fasta_file_clean = f\"{'_'.join(str(virus).split(' '))}_aligned_headers_adjusted.afa\"\n",
        "update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean)\n",
        "\n",
        "log_message(\"4/5 Reformatting complete.\")\n",
        "\n",
        "# Run Delphy\n",
        "log_message(\"5/5 Running Delphy...\")\n",
        "\n",
        "# Download delphy binary\n",
        "!wget https://github.com/broadinstitute/delphy/releases/download/0.9995/delphy-ubuntu-x86_64\n",
        "\n",
        "# Give permissions\n",
        "!chmod u+x ./delphy-ubuntu-x86_64\n",
        "\n",
        "beast_log_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_log.txt\"\n",
        "delphy_beast_tree_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_tree.nwk\"\n",
        "dphy_out = f\"{'_'.join(str(virus).split(' '))}_delphy_out.dphy\"\n",
        "\n",
        "# Delphy settings\n",
        "# TO-DO allow users to modify Delphy settings\n",
        "threads = 2\n",
        "delphy_steps = 10000000\n",
        "mutation_rate = 1e-3    # If you want to fix the mutation rate (if not, set to None)\n",
        "log_every = int(delphy_steps / 100)\n",
        "tree_every = int(delphy_steps / 100)\n",
        "\n",
        "if mutation_rate:\n",
        "  !./delphy-ubuntu-x86_64 \\\n",
        "    --v0-fix-mutation-rate \\\n",
        "    --v0-init-mutation-rate $mutation_rate \\\n",
        "    --v0-log-every $log_every \\\n",
        "    --v0-tree-every $tree_every \\\n",
        "    --v0-threads $threads \\\n",
        "    --v0-steps $delphy_steps \\\n",
        "    --v0-in-fasta $aligned_fasta_file_clean \\\n",
        "    --v0-out-log-file $beast_log_out \\\n",
        "    --v0-out-trees-file $delphy_beast_tree_out \\\n",
        "    --v0-out-delphy-file $dphy_out\n",
        "else:\n",
        "    !./delphy-ubuntu-x86_64 \\\n",
        "      --v0-log-every $log_every \\\n",
        "      --v0-tree-every $tree_every \\\n",
        "      --v0-threads $threads \\\n",
        "      --v0-steps $delphy_steps \\\n",
        "      --v0-in-fasta $aligned_fasta_file_clean \\\n",
        "      --v0-out-log-file $beast_log_out \\\n",
        "      --v0-out-trees-file $delphy_beast_tree_out \\\n",
        "      --v0-out-delphy-file $dphy_out\n",
        "\n",
        "# TO-DO: Only print message below if no error occured above\n",
        "\n",
        "# Display a message when done\n",
        "display(HTML(\"\"\"\n",
        "<h1>All done! 🎉</h1>\n",
        "<h2>To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.</h2>\n",
        "<h2>To further visualize your Delphy output, upload the <code>.dphy</code> (and <code>metadata.csv</code>) file(s) to <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a></h2>\n",
        "\"\"\"))\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HuVr7cb0x5VF",
        "outputId": "4744f68a-7e6d-43a0-869b-e5e0a274ad89",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>1/5 Installing software...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.0/34.0 MB\u001b[0m \u001b[31m16.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m15.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for gget (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>1/5 Software installation complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>2/5 Download data from NCBI Virus... This might take a few minutes depending on the internet connection and how busy the NCBI server is.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "INFO:gget.utils:119 sequences passed the provided filters.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>2/5 Data download from NCBI Virus complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package fonts-lato.\n",
            "(Reading database ... 124926 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-lato_2.0-2.1_all.deb ...\n",
            "Unpacking fonts-lato (2.0-2.1) ...\n",
            "Selecting previously unselected package netbase.\n",
            "Preparing to unpack .../01-netbase_6.3_all.deb ...\n",
            "Unpacking netbase (6.3) ...\n",
            "Selecting previously unselected package libclone-perl.\n",
            "Preparing to unpack .../02-libclone-perl_0.45-1build3_amd64.deb ...\n",
            "Unpacking libclone-perl (0.45-1build3) ...\n",
            "Selecting previously unselected package libdata-dump-perl.\n",
            "Preparing to unpack .../03-libdata-dump-perl_1.25-1_all.deb ...\n",
            "Unpacking libdata-dump-perl (1.25-1) ...\n",
            "Selecting previously unselected package libencode-locale-perl.\n",
            "Preparing to unpack .../04-libencode-locale-perl_1.05-1.1_all.deb ...\n",
            "Unpacking libencode-locale-perl (1.05-1.1) ...\n",
            "Selecting previously unselected package libhttp-date-perl.\n",
            "Preparing to unpack .../05-libhttp-date-perl_6.05-1_all.deb ...\n",
            "Unpacking libhttp-date-perl (6.05-1) ...\n",
            "Selecting previously unselected package libfile-listing-perl.\n",
            "Preparing to unpack .../06-libfile-listing-perl_6.14-1_all.deb ...\n",
            "Unpacking libfile-listing-perl (6.14-1) ...\n",
            "Selecting previously unselected package libfont-afm-perl.\n",
            "Preparing to unpack .../07-libfont-afm-perl_1.20-3_all.deb ...\n",
            "Unpacking libfont-afm-perl (1.20-3) ...\n",
            "Selecting previously unselected package libhtml-tagset-perl.\n",
            "Preparing to unpack .../08-libhtml-tagset-perl_3.20-4_all.deb ...\n",
            "Unpacking libhtml-tagset-perl (3.20-4) ...\n",
            "Selecting previously unselected package liburi-perl.\n",
            "Preparing to unpack .../09-liburi-perl_5.10-1_all.deb ...\n",
            "Unpacking liburi-perl (5.10-1) ...\n",
            "Selecting previously unselected package libhtml-parser-perl:amd64.\n",
            "Preparing to unpack .../10-libhtml-parser-perl_3.76-1build2_amd64.deb ...\n",
            "Unpacking libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Selecting previously unselected package libio-html-perl.\n",
            "Preparing to unpack .../11-libio-html-perl_1.004-2_all.deb ...\n",
            "Unpacking libio-html-perl (1.004-2) ...\n",
            "Selecting previously unselected package liblwp-mediatypes-perl.\n",
            "Preparing to unpack .../12-liblwp-mediatypes-perl_6.04-1_all.deb ...\n",
            "Unpacking liblwp-mediatypes-perl (6.04-1) ...\n",
            "Selecting previously unselected package libhttp-message-perl.\n",
            "Preparing to unpack .../13-libhttp-message-perl_6.36-1_all.deb ...\n",
            "Unpacking libhttp-message-perl (6.36-1) ...\n",
            "Selecting previously unselected package libhtml-form-perl.\n",
            "Preparing to unpack .../14-libhtml-form-perl_6.07-1_all.deb ...\n",
            "Unpacking libhtml-form-perl (6.07-1) ...\n",
            "Selecting previously unselected package libhtml-tree-perl.\n",
            "Preparing to unpack .../15-libhtml-tree-perl_5.07-2_all.deb ...\n",
            "Unpacking libhtml-tree-perl (5.07-2) ...\n",
            "Selecting previously unselected package libhtml-format-perl.\n",
            "Preparing to unpack .../16-libhtml-format-perl_2.12-1.1_all.deb ...\n",
            "Unpacking libhtml-format-perl (2.12-1.1) ...\n",
            "Selecting previously unselected package libhttp-cookies-perl.\n",
            "Preparing to unpack .../17-libhttp-cookies-perl_6.10-1_all.deb ...\n",
            "Unpacking libhttp-cookies-perl (6.10-1) ...\n",
            "Selecting previously unselected package libhttp-daemon-perl.\n",
            "Preparing to unpack .../18-libhttp-daemon-perl_6.13-1ubuntu0.1_all.deb ...\n",
            "Unpacking libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libhttp-negotiate-perl.\n",
            "Preparing to unpack .../19-libhttp-negotiate-perl_6.01-1_all.deb ...\n",
            "Unpacking libhttp-negotiate-perl (6.01-1) ...\n",
            "Selecting previously unselected package perl-openssl-defaults:amd64.\n",
            "Preparing to unpack .../20-perl-openssl-defaults_5build2_amd64.deb ...\n",
            "Unpacking perl-openssl-defaults:amd64 (5build2) ...\n",
            "Selecting previously unselected package libnet-ssleay-perl:amd64.\n",
            "Preparing to unpack .../21-libnet-ssleay-perl_1.92-1build2_amd64.deb ...\n",
            "Unpacking libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Selecting previously unselected package libio-socket-ssl-perl.\n",
            "Preparing to unpack .../22-libio-socket-ssl-perl_2.074-2_all.deb ...\n",
            "Unpacking libio-socket-ssl-perl (2.074-2) ...\n",
            "Selecting previously unselected package libnet-http-perl.\n",
            "Preparing to unpack .../23-libnet-http-perl_6.22-1_all.deb ...\n",
            "Unpacking libnet-http-perl (6.22-1) ...\n",
            "Selecting previously unselected package libtry-tiny-perl.\n",
            "Preparing to unpack .../24-libtry-tiny-perl_0.31-1_all.deb ...\n",
            "Unpacking libtry-tiny-perl (0.31-1) ...\n",
            "Selecting previously unselected package libwww-robotrules-perl.\n",
            "Preparing to unpack .../25-libwww-robotrules-perl_6.02-1_all.deb ...\n",
            "Unpacking libwww-robotrules-perl (6.02-1) ...\n",
            "Selecting previously unselected package libwww-perl.\n",
            "Preparing to unpack .../26-libwww-perl_6.61-1_all.deb ...\n",
            "Unpacking libwww-perl (6.61-1) ...\n",
            "Selecting previously unselected package liblwp-protocol-https-perl.\n",
            "Preparing to unpack .../27-liblwp-protocol-https-perl_6.10-1_all.deb ...\n",
            "Unpacking liblwp-protocol-https-perl (6.10-1) ...\n",
            "Selecting previously unselected package libnet-smtp-ssl-perl.\n",
            "Preparing to unpack .../28-libnet-smtp-ssl-perl_1.04-1_all.deb ...\n",
            "Unpacking libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Selecting previously unselected package libmailtools-perl.\n",
            "Preparing to unpack .../29-libmailtools-perl_2.21-1_all.deb ...\n",
            "Unpacking libmailtools-perl (2.21-1) ...\n",
            "Selecting previously unselected package rubygems-integration.\n",
            "Preparing to unpack .../30-rubygems-integration_1.18_all.deb ...\n",
            "Unpacking rubygems-integration (1.18) ...\n",
            "Selecting previously unselected package ruby3.0.\n",
            "Preparing to unpack .../31-ruby3.0_3.0.2-7ubuntu2.8_amd64.deb ...\n",
            "Unpacking ruby3.0 (3.0.2-7ubuntu2.8) ...\n",
            "Selecting previously unselected package ruby-rubygems.\n",
            "Preparing to unpack .../32-ruby-rubygems_3.3.5-2_all.deb ...\n",
            "Unpacking ruby-rubygems (3.3.5-2) ...\n",
            "Selecting previously unselected package ruby.\n",
            "Preparing to unpack .../33-ruby_1%3a3.0~exp1_amd64.deb ...\n",
            "Unpacking ruby (1:3.0~exp1) ...\n",
            "Selecting previously unselected package rake.\n",
            "Preparing to unpack .../34-rake_13.0.6-2_all.deb ...\n",
            "Unpacking rake (13.0.6-2) ...\n",
            "Selecting previously unselected package ruby-net-telnet.\n",
            "Preparing to unpack .../35-ruby-net-telnet_0.1.1-2_all.deb ...\n",
            "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
            "Selecting previously unselected package ruby-webrick.\n",
            "Preparing to unpack .../36-ruby-webrick_1.7.0-3ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Selecting previously unselected package ruby-xmlrpc.\n",
            "Preparing to unpack .../37-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libruby3.0:amd64.\n",
            "Preparing to unpack .../38-libruby3.0_3.0.2-7ubuntu2.8_amd64.deb ...\n",
            "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.8) ...\n",
            "Selecting previously unselected package lynx-common.\n",
            "Preparing to unpack .../39-lynx-common_2.9.0dev.10-1_all.deb ...\n",
            "Unpacking lynx-common (2.9.0dev.10-1) ...\n",
            "Selecting previously unselected package mafft.\n",
            "Preparing to unpack .../40-mafft_7.490-1_amd64.deb ...\n",
            "Unpacking mafft (7.490-1) ...\n",
            "Selecting previously unselected package libauthen-sasl-perl.\n",
            "Preparing to unpack .../41-libauthen-sasl-perl_2.1600-1.1_all.deb ...\n",
            "Unpacking libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Selecting previously unselected package lynx.\n",
            "Preparing to unpack .../42-lynx_2.9.0dev.10-1_amd64.deb ...\n",
            "Unpacking lynx (2.9.0dev.10-1) ...\n",
            "Setting up libhttp-date-perl (6.05-1) ...\n",
            "Setting up fonts-lato (2.0-2.1) ...\n",
            "Setting up libfile-listing-perl (6.14-1) ...\n",
            "Setting up libfont-afm-perl (1.20-3) ...\n",
            "Setting up mafft (7.490-1) ...\n",
            "Setting up libclone-perl (0.45-1build3) ...\n",
            "Setting up libhtml-tagset-perl (3.20-4) ...\n",
            "Setting up libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Setting up liblwp-mediatypes-perl (6.04-1) ...\n",
            "Setting up libtry-tiny-perl (0.31-1) ...\n",
            "Setting up perl-openssl-defaults:amd64 (5build2) ...\n",
            "Setting up libencode-locale-perl (1.05-1.1) ...\n",
            "Setting up rubygems-integration (1.18) ...\n",
            "Setting up libdata-dump-perl (1.25-1) ...\n",
            "Setting up ruby-net-telnet (0.1.1-2) ...\n",
            "Setting up libio-html-perl (1.004-2) ...\n",
            "Setting up lynx-common (2.9.0dev.10-1) ...\n",
            "Setting up ruby-webrick (1.7.0-3ubuntu0.1) ...\n",
            "Setting up netbase (6.3) ...\n",
            "Setting up lynx (2.9.0dev.10-1) ...\n",
            "update-alternatives: using /usr/bin/lynx to provide /usr/bin/www-browser (www-browser) in auto mode\n",
            "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Setting up liburi-perl (5.10-1) ...\n",
            "Setting up libhttp-message-perl (6.36-1) ...\n",
            "Setting up libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Setting up libhttp-negotiate-perl (6.01-1) ...\n",
            "Setting up libhttp-cookies-perl (6.10-1) ...\n",
            "Setting up libnet-http-perl (6.22-1) ...\n",
            "Setting up libwww-robotrules-perl (6.02-1) ...\n",
            "Setting up libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Setting up libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Setting up libio-socket-ssl-perl (2.074-2) ...\n",
            "Setting up libhtml-form-perl (6.07-1) ...\n",
            "Setting up libhtml-tree-perl (5.07-2) ...\n",
            "Setting up libhtml-format-perl (2.12-1.1) ...\n",
            "Setting up libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Setting up libmailtools-perl (2.21-1) ...\n",
            "Setting up libwww-perl (6.61-1) ...\n",
            "Setting up ruby3.0 (3.0.2-7ubuntu2.8) ...\n",
            "Setting up liblwp-protocol-https-perl (6.10-1) ...\n",
            "Setting up ruby (1:3.0~exp1) ...\n",
            "Setting up rake (13.0.6-2) ...\n",
            "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.8) ...\n",
            "Setting up ruby-rubygems (3.3.5-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for mailcap (3.70+nmu1ubuntu1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>3/5 MSA complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>4/5 Reformatting sequence files to match Delphy format...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>4/5 Reformatting complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>5/5 Running Delphy...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-02-07 17:30:47--  https://github.com/broadinstitute/delphy/releases/download/0.9995/delphy-ubuntu-x86_64\n",
            "Resolving github.com (github.com)... 140.82.113.3\n",
            "Connecting to github.com (github.com)|140.82.113.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/804287442/5e0b7be1-42ce-451f-a096-167441c1a12a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250207%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250207T173047Z&X-Amz-Expires=300&X-Amz-Signature=e9b2318873356ebc0bbc5cc1551882a762ef0ba0bdf1c8824a0d85930e54afe7&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Ddelphy-ubuntu-x86_64&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-02-07 17:30:47--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/804287442/5e0b7be1-42ce-451f-a096-167441c1a12a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250207%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250207T173047Z&X-Amz-Expires=300&X-Amz-Signature=e9b2318873356ebc0bbc5cc1551882a762ef0ba0bdf1c8824a0d85930e54afe7&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Ddelphy-ubuntu-x86_64&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 2165232 (2.1M) [application/octet-stream]\n",
            "Saving to: ‘delphy-ubuntu-x86_64’\n",
            "\n",
            "\rdelphy-ubuntu-x86_6   0%[                    ]       0  --.-KB/s               \rdelphy-ubuntu-x86_6 100%[===================>]   2.06M  --.-KB/s    in 0.01s   \n",
            "\n",
            "2025-02-07 17:30:47 (140 MB/s) - ‘delphy-ubuntu-x86_64’ saved [2165232/2165232]\n",
            "\n",
            "# Seed: 1400166009\n",
            "Reading fasta file sars-cov-2_aligned_headers_adjusted.afa\n",
            "Building rough initial tree...\n",
            "Warning: ambiguous state Y at site 23646 changed to N\n",
            "Warning: ambiguous state Y at site 3565 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state Y at site 3565 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state Y at site 3565 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state Y at site 3565 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state Y at site 3565 changed to N\n",
            "Warning: ambiguous state R at site 13427 changed to N\n",
            "Warning: ambiguous state R at site 26604 changed to N\n",
            "Warning: ambiguous state Y at site 3020 changed to N\n",
            "Warning: ambiguous state W at site 26524 changed to N\n",
            "Warning: ambiguous state R at site 13427 changed to N\n",
            "Warning: ambiguous state R at site 13427 changed to N\n",
            "Attaching tip 2 to Candidate_region{branch=1, mut_idx=15, 1842.12<t<=1846, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 3 to Candidate_region{branch=119, mut_idx=0, -1.79769e+308<t<=1613, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 4 to Candidate_region{branch=0, mut_idx=1, 1615.74<t<=1640.46, min_muts=27, W_over_Wmax=0}\n",
            "Attaching tip 5 to Candidate_region{branch=4, mut_idx=27, 1842.77<t<=1843, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 6 to Candidate_region{branch=120, mut_idx=11, 1809.41<t<=1819.41, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 7 to Candidate_region{branch=3, mut_idx=16, 1844.39<t<=1845, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 8 to Candidate_region{branch=121, mut_idx=0, -1.79769e+308<t<=1612, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 9 to Candidate_region{branch=6, mut_idx=4, 1826.79<t<=1835.5, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 10 to Candidate_region{branch=126, mut_idx=0, -1.79769e+308<t<=1611, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 11 to Candidate_region{branch=6, mut_idx=0, 1827.96<t<=1835.5, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 12 to Candidate_region{branch=10, mut_idx=3, 1664.36<t<=1673.23, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 13 to Candidate_region{branch=128, mut_idx=0, -1.79769e+308<t<=1610, min_muts=20, W_over_Wmax=0}\n",
            "Attaching tip 14 to Candidate_region{branch=12, mut_idx=0, 1670.89<t<=1716.07, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 15 to Candidate_region{branch=125, mut_idx=15, 1838.22<t<=1842, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 16 to Candidate_region{branch=129, mut_idx=0, 1827.96<t<=1831.77, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 17 to Candidate_region{branch=12, mut_idx=0, 1688.37<t<=1716.07, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 18 to Candidate_region{branch=133, mut_idx=15, 1838.22<t<=1841, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 19 to Candidate_region{branch=17, mut_idx=0, 1701.47<t<=1729.88, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 20 to Candidate_region{branch=129, mut_idx=0, 1828.76<t<=1831.77, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 21 to Candidate_region{branch=20, mut_idx=5, 1839.5<t<=1841, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 22 to Candidate_region{branch=139, mut_idx=2, 1833.19<t<=1836.73, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 23 to Candidate_region{branch=15, mut_idx=9, 1841.57<t<=1841.71, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 24 to Candidate_region{branch=14, mut_idx=1, 1697.91<t<=1707.67, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 25 to Candidate_region{branch=139, mut_idx=0, 1834.1<t<=1836.73, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 26 to Candidate_region{branch=24, mut_idx=1, 1718.67<t<=1737.12, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 27 to Candidate_region{branch=132, mut_idx=0, 1670.89<t<=1688.37, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 28 to Candidate_region{branch=23, mut_idx=2, 1841.76<t<=1841.82, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 29 to Candidate_region{branch=28, mut_idx=0, 1841.81<t<=1843.95, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 30 to Candidate_region{branch=29, mut_idx=0, 1841.92<t<=1843.79, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 31 to Candidate_region{branch=131, mut_idx=0, -1.79769e+308<t<=1580, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 32 to Candidate_region{branch=145, mut_idx=0, 1670.89<t<=1683.25, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 33 to Candidate_region{branch=12, mut_idx=0, 1701.47<t<=1716.07, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 34 to Candidate_region{branch=25, mut_idx=0, 1835.44<t<=1838.99, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 35 to Candidate_region{branch=28, mut_idx=3, 1844.91<t<=1845, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 36 to Candidate_region{branch=25, mut_idx=2, 1848.01<t<=1850, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 37 to Candidate_region{branch=18, mut_idx=4, 1840.69<t<=1840.71, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 38 to Candidate_region{branch=153, mut_idx=0, 1841.92<t<=1843.95, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 39 to Candidate_region{branch=154, mut_idx=0, 1838.5<t<=1838.99, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 40 to Candidate_region{branch=132, mut_idx=0, 1683.25<t<=1688.37, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 41 to Candidate_region{branch=147, mut_idx=0, 1841.81<t<=1841.92, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 42 to Candidate_region{branch=152, mut_idx=0, 1835.44<t<=1838.5, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 43 to Candidate_region{branch=141, mut_idx=8, 1841.56<t<=1841.57, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 44 to Candidate_region{branch=41, mut_idx=1, 1841.91<t<=1841.91, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 45 to Candidate_region{branch=44, mut_idx=0, 1841.91<t<=1845.24, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 46 to Candidate_region{branch=137, mut_idx=0, 1701.47<t<=1724.44, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 47 to Candidate_region{branch=14, mut_idx=8, 1814.99<t<=1821.89, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 48 to Candidate_region{branch=149, mut_idx=0, -1.79769e+308<t<=1579, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 49 to Candidate_region{branch=42, mut_idx=0, 1837.81<t<=1838.25, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 50 to Candidate_region{branch=49, mut_idx=2, 1840.97<t<=1841.83, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 51 to Candidate_region{branch=37, mut_idx=0, 1840.7<t<=1840.74, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 52 to Candidate_region{branch=48, mut_idx=13, 1805.54<t<=1823.28, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 53 to Candidate_region{branch=44, mut_idx=0, 1842.09<t<=1843, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 54 to Candidate_region{branch=45, mut_idx=2, 1843.15<t<=1843.37, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 55 to Candidate_region{branch=26, mut_idx=0, 1727.82<t<=1737.15, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 56 to Candidate_region{branch=54, mut_idx=1, 1843.37<t<=1843.8, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 57 to Candidate_region{branch=47, mut_idx=3, 1825.22<t<=1827.99, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 58 to Candidate_region{branch=163, mut_idx=0, 1841.91<t<=1842, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 59 to Candidate_region{branch=174, mut_idx=1, 1843.37<t<=1843.66, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 60 to Candidate_region{branch=59, mut_idx=0, 1843.39<t<=1844.67, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 61 to Candidate_region{branch=23, mut_idx=7, 1841.93<t<=1841.94, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 62 to Candidate_region{branch=22, mut_idx=0, 1834.1<t<=1835.84, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 63 to Candidate_region{branch=60, mut_idx=0, 1843.84<t<=1844.77, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 64 to Candidate_region{branch=13, mut_idx=9, 1687.53<t<=1713.38, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 65 to Candidate_region{branch=130, mut_idx=1, 1624.53<t<=1650.67, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 66 to Candidate_region{branch=45, mut_idx=5, 1844.77<t<=1846, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 67 to Candidate_region{branch=178, mut_idx=0, 1843.39<t<=1843.84, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 68 to Candidate_region{branch=160, mut_idx=0, 1835.44<t<=1837.81, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 69 to Candidate_region{branch=56, mut_idx=1, 1843.89<t<=1844.31, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 70 to Candidate_region{branch=166, mut_idx=0, -1.79769e+308<t<=1578, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 71 to Candidate_region{branch=184, mut_idx=2, 1843.48<t<=1844.07, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 72 to Candidate_region{branch=55, mut_idx=0, 1733.19<t<=1752.87, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 73 to Candidate_region{branch=185, mut_idx=0, 1843.39<t<=1843.81, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 74 to Candidate_region{branch=139, mut_idx=3, 1839.5<t<=1840.01, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 75 to Candidate_region{branch=191, mut_idx=0, 1843.39<t<=1843.7, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 76 to Candidate_region{branch=186, mut_idx=0, 1835.44<t<=1836.84, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 77 to Candidate_region{branch=152, mut_idx=0, 1837.81<t<=1838.5, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 78 to Candidate_region{branch=44, mut_idx=4, 1846.91<t<=1849, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 79 to Candidate_region{branch=168, mut_idx=0, 1837.91<t<=1840.46, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 80 to Candidate_region{branch=75, mut_idx=0, 1843.57<t<=1845.26, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 81 to Candidate_region{branch=79, mut_idx=0, 1840.38<t<=1841.72, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 82 to Candidate_region{branch=46, mut_idx=0, 1716.53<t<=1721.23, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 83 to Candidate_region{branch=82, mut_idx=1, 1738.76<t<=1739.76, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 84 to Candidate_region{branch=75, mut_idx=0, 1843.92<t<=1845.26, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 85 to Candidate_region{branch=62, mut_idx=3, 1843.82<t<=1846, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 86 to Candidate_region{branch=27, mut_idx=2, 1689.39<t<=1716.52, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 87 to Candidate_region{branch=192, mut_idx=2, 1837.51<t<=1839.5, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 88 to Candidate_region{branch=64, mut_idx=7, 1733.45<t<=1763.18, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 89 to Candidate_region{branch=164, mut_idx=0, 1701.47<t<=1716.53, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 90 to Candidate_region{branch=77, mut_idx=4, 1842.22<t<=1842.41, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 91 to Candidate_region{branch=88, mut_idx=12, 1839.61<t<=1852, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 92 to Candidate_region{branch=87, mut_idx=0, 1839.07<t<=1844.63, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 93 to Candidate_region{branch=22, mut_idx=6, 1841.96<t<=1842, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 94 to Candidate_region{branch=52, mut_idx=8, 1841.3<t<=1842, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 95 to Candidate_region{branch=6, mut_idx=4, 1840.96<t<=1847, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 96 to Candidate_region{branch=64, mut_idx=0, 1741.14<t<=1763.18, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 97 to Candidate_region{branch=203, mut_idx=0, 1834.56<t<=1841, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 98 to Candidate_region{branch=169, mut_idx=0, 1840.7<t<=1840.74, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 99 to Candidate_region{branch=135, mut_idx=0, 1688.37<t<=1701.47, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 100 to Candidate_region{branch=155, mut_idx=4, 1840.69<t<=1840.7, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 101 to Candidate_region{branch=216, mut_idx=0, 1840.7<t<=1840.73, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 102 to Candidate_region{branch=190, mut_idx=0, 1733.19<t<=1744.01, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 103 to Candidate_region{branch=26, mut_idx=0, 1733.19<t<=1737.15, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 104 to Candidate_region{branch=102, mut_idx=10, 1826.89<t<=1839.35, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 105 to Candidate_region{branch=39, mut_idx=2, 1840.49<t<=1842, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 106 to Candidate_region{branch=58, mut_idx=1, 1841.98<t<=1841.98, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 107 to Candidate_region{branch=224, mut_idx=1, 1841.98<t<=1841.98, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 108 to Candidate_region{branch=197, mut_idx=0, 1837.91<t<=1840.38, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 109 to Candidate_region{branch=170, mut_idx=3, 1622.87<t<=1671.68, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 110 to Candidate_region{branch=72, mut_idx=4, 1780.08<t<=1810.4, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 111 to Candidate_region{branch=219, mut_idx=0, 1840.7<t<=1840.72, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 112 to Candidate_region{branch=83, mut_idx=0, 1739.57<t<=1740.6, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 113 to Candidate_region{branch=216, mut_idx=0, 1840.72<t<=1840.73, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 114 to Candidate_region{branch=169, mut_idx=0, 1840.73<t<=1840.74, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 115 to Candidate_region{branch=98, mut_idx=0, 1840.73<t<=1840.74, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 116 to Candidate_region{branch=203, mut_idx=0, 1838.04<t<=1841, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 117 to Candidate_region{branch=115, mut_idx=1, 1840.74<t<=1840.83, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 118 to Candidate_region{branch=60, mut_idx=1, 1844.77<t<=1845, min_muts=3, W_over_Wmax=0}\n",
            "Built initial tree with 119 tips\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "# Parallelism: 2\n",
            "0.0 M steps/s, Step 0, log_posterior = -50700.44, log_G = -55797.41, log_coal = -1582.06, log_other_priors = 6679.03, num_muts = 975, T = 22004.90, t_MRCA = 2023-06-20, n0 = 1000.0, g = 0.0000, mu = 1 * 10^-3 subst / site / year, M_ts = 760, M_tv = 215, kappa = 1, pi = [0.3, 0.18, 0.2, 0.32], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [12, 98, 17, 21, 4, 432, 59, 20, 81, 15, 171, 45]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.4 M steps/s, Step 100000, log_posterior = -46650.58, log_G = -52276.76, log_coal = -1047.88, log_other_priors = 6674.06, num_muts = 795, T = 14474.61, t_MRCA = 2023-11-26, n0 = 10828.3, g = 0.0103, mu = 1 * 10^-3 subst / site / year, M_ts = 636, M_tv = 159, kappa = 9.63, pi = [0.3, 0.18, 0.19, 0.34], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [10, 67, 17, 20, 3, 398, 53, 15, 65, 14, 118, 15]\n",
            "0.5 M steps/s, Step 200000, log_posterior = -45864.20, log_G = -51520.43, log_coal = -1017.99, log_other_priors = 6674.22, num_muts = 757, T = 11623.45, t_MRCA = 2024-01-21, n0 = 10770.6, g = 0.0154, mu = 1 * 10^-3 subst / site / year, M_ts = 609, M_tv = 148, kappa = 8.57, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5855, 9616], num_muts_ab counts: [11, 64, 14, 20, 2, 392, 50, 14, 62, 14, 103, 11]\n",
            "0.5 M steps/s, Step 300000, log_posterior = -45641.12, log_G = -51306.96, log_coal = -1008.08, log_other_priors = 6673.92, num_muts = 745, T = 10973.98, t_MRCA = 2024-04-12, n0 = 13759.4, g = 0.0195, mu = 1 * 10^-3 subst / site / year, M_ts = 600, M_tv = 145, kappa = 8.6, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5855, 9616], num_muts_ab counts: [9, 62, 15, 19, 2, 388, 47, 14, 62, 13, 103, 11]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 0.867463 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.5 M steps/s, Step 400000, log_posterior = -45491.86, log_G = -51171.48, log_coal = -994.07, log_other_priors = 6673.70, num_muts = 738, T = 10441.53, t_MRCA = 2024-05-06, n0 = 16463.6, g = 0.0238, mu = 1 * 10^-3 subst / site / year, M_ts = 593, M_tv = 145, kappa = 8.54, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 61, 15, 19, 2, 386, 47, 14, 62, 13, 99, 11]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 0.757479 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.5 M steps/s, Step 500000, log_posterior = -45408.34, log_G = -51091.18, log_coal = -990.61, log_other_priors = 6673.45, num_muts = 734, T = 10097.05, t_MRCA = 2024-06-10, n0 = 18674.7, g = 0.0279, mu = 1 * 10^-3 subst / site / year, M_ts = 590, M_tv = 144, kappa = 8.91, pi = [0.3, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [8, 61, 15, 19, 2, 383, 47, 15, 60, 13, 99, 12]\n",
            "0.5 M steps/s, Step 600000, log_posterior = -45371.67, log_G = -51061.14, log_coal = -984.00, log_other_priors = 6673.48, num_muts = 733, T = 9874.41, t_MRCA = 2024-06-12, n0 = 23694.0, g = 0.0298, mu = 1 * 10^-3 subst / site / year, M_ts = 589, M_tv = 144, kappa = 7.51, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [9, 61, 15, 19, 2, 381, 47, 14, 60, 13, 100, 12]\n",
            "0.5 M steps/s, Step 700000, log_posterior = -45253.69, log_G = -50946.96, log_coal = -979.73, log_other_priors = 6673.00, num_muts = 726, T = 9625.98, t_MRCA = 2024-06-08, n0 = 27425.4, g = 0.0317, mu = 1 * 10^-3 subst / site / year, M_ts = 583, M_tv = 143, kappa = 9, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [9, 61, 14, 19, 2, 377, 47, 14, 60, 13, 98, 12]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 0.655731 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.5 M steps/s, Step 800000, log_posterior = -45090.31, log_G = -50781.37, log_coal = -982.79, log_other_priors = 6673.84, num_muts = 716, T = 9307.92, t_MRCA = 2024-06-02, n0 = 11601.5, g = 0.0238, mu = 1 * 10^-3 subst / site / year, M_ts = 578, M_tv = 138, kappa = 9.59, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5459, 5854, 9617], num_muts_ab counts: [10, 60, 13, 18, 2, 374, 47, 13, 60, 12, 97, 10]\n",
            "0.3 M steps/s, Step 900000, log_posterior = -45030.89, log_G = -50730.70, log_coal = -973.13, log_other_priors = 6672.94, num_muts = 713, T = 9111.57, t_MRCA = 2024-06-30, n0 = 23469.8, g = 0.0329, mu = 1 * 10^-3 subst / site / year, M_ts = 575, M_tv = 138, kappa = 10.1, pi = [0.29, 0.18, 0.19, 0.34], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 60, 13, 18, 2, 374, 47, 14, 61, 12, 94, 9]\n",
            "0.3 M steps/s, Step 1000000, log_posterior = -45049.43, log_G = -50755.58, log_coal = -966.70, log_other_priors = 6672.86, num_muts = 713, T = 9469.83, t_MRCA = 2024-07-06, n0 = 31654.6, g = 0.0364, mu = 1 * 10^-3 subst / site / year, M_ts = 575, M_tv = 138, kappa = 8.72, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 60, 13, 18, 2, 374, 47, 14, 61, 12, 94, 9]\n",
            "0.3 M steps/s, Step 1100000, log_posterior = -45037.87, log_G = -50739.79, log_coal = -970.69, log_other_priors = 6672.61, num_muts = 711, T = 9599.00, t_MRCA = 2024-06-28, n0 = 37220.6, g = 0.0351, mu = 1 * 10^-3 subst / site / year, M_ts = 573, M_tv = 138, kappa = 9.22, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5461, 5855, 9614], num_muts_ab counts: [10, 59, 13, 17, 2, 375, 47, 14, 60, 12, 92, 10]\n",
            "0.3 M steps/s, Step 1200000, log_posterior = -45013.33, log_G = -50712.78, log_coal = -973.73, log_other_priors = 6673.19, num_muts = 711, T = 9258.13, t_MRCA = 2024-07-04, n0 = 28669.2, g = 0.0360, mu = 1 * 10^-3 subst / site / year, M_ts = 573, M_tv = 138, kappa = 7.63, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5461, 5855, 9614], num_muts_ab counts: [10, 59, 13, 17, 2, 375, 47, 14, 60, 12, 92, 10]\n",
            "0.3 M steps/s, Step 1300000, log_posterior = -45012.38, log_G = -50716.72, log_coal = -968.58, log_other_priors = 6672.93, num_muts = 711, T = 9311.06, t_MRCA = 2024-07-02, n0 = 29670.0, g = 0.0355, mu = 1 * 10^-3 subst / site / year, M_ts = 573, M_tv = 138, kappa = 8.73, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5464, 5853, 9614], num_muts_ab counts: [9, 60, 13, 18, 4, 376, 47, 12, 60, 12, 90, 10]\n",
            "0.3 M steps/s, Step 1400000, log_posterior = -44990.45, log_G = -50705.86, log_coal = -957.02, log_other_priors = 6672.42, num_muts = 711, T = 9149.00, t_MRCA = 2024-07-17, n0 = 48371.5, g = 0.0427, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 139, kappa = 8.42, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5463, 5854, 9614], num_muts_ab counts: [9, 59, 13, 18, 4, 375, 47, 13, 60, 12, 91, 10]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 0.570566 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.3 M steps/s, Step 1500000, log_posterior = -44940.01, log_G = -50652.08, log_coal = -960.91, log_other_priors = 6672.98, num_muts = 708, T = 9005.18, t_MRCA = 2024-06-21, n0 = 28822.6, g = 0.0376, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.5, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 59, 13, 18, 2, 374, 47, 14, 60, 12, 90, 10]\n",
            "0.3 M steps/s, Step 1600000, log_posterior = -44965.82, log_G = -50677.20, log_coal = -960.82, log_other_priors = 6672.20, num_muts = 708, T = 9313.60, t_MRCA = 2024-07-09, n0 = 59950.1, g = 0.0443, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.33, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 59, 13, 18, 2, 374, 47, 14, 60, 12, 90, 10]\n",
            "0.3 M steps/s, Step 1700000, log_posterior = -44987.38, log_G = -50692.46, log_coal = -967.76, log_other_priors = 6672.85, num_muts = 708, T = 9499.21, t_MRCA = 2024-06-25, n0 = 33243.2, g = 0.0365, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.52, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5464, 5854, 9613], num_muts_ab counts: [9, 59, 13, 18, 4, 374, 47, 12, 61, 12, 90, 9]\n",
            "0.3 M steps/s, Step 1800000, log_posterior = -45003.51, log_G = -50698.35, log_coal = -977.95, log_other_priors = 6672.78, num_muts = 708, T = 9543.18, t_MRCA = 2024-06-17, n0 = 35343.0, g = 0.0337, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.7, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 59, 13, 18, 2, 374, 47, 14, 60, 12, 90, 10]\n",
            "0.4 M steps/s, Step 1900000, log_posterior = -45020.38, log_G = -50721.14, log_coal = -971.72, log_other_priors = 6672.49, num_muts = 709, T = 9705.57, t_MRCA = 2024-07-11, n0 = 42023.0, g = 0.0376, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 138, kappa = 9.08, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 376, 47, 14, 60, 12, 90, 10]\n",
            "0.5 M steps/s, Step 2000000, log_posterior = -44968.95, log_G = -50670.57, log_coal = -971.74, log_other_priors = 6673.35, num_muts = 709, T = 9073.75, t_MRCA = 2024-07-09, n0 = 16232.6, g = 0.0310, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 138, kappa = 9.95, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 376, 47, 14, 60, 12, 90, 10]\n",
            "0.5 M steps/s, Step 2100000, log_posterior = -44982.61, log_G = -50681.70, log_coal = -974.46, log_other_priors = 6673.55, num_muts = 708, T = 9368.10, t_MRCA = 2024-06-18, n0 = 19493.0, g = 0.0306, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.05, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 375, 47, 14, 60, 12, 90, 10]\n",
            "0.5 M steps/s, Step 2200000, log_posterior = -44998.75, log_G = -50695.95, log_coal = -975.42, log_other_priors = 6672.63, num_muts = 708, T = 9563.08, t_MRCA = 2024-06-23, n0 = 33941.2, g = 0.0351, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.64, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5464, 5854, 9613], num_muts_ab counts: [9, 58, 13, 18, 4, 376, 47, 12, 61, 12, 90, 8]\n",
            "0.5 M steps/s, Step 2300000, log_posterior = -45048.86, log_G = -50740.05, log_coal = -981.66, log_other_priors = 6672.86, num_muts = 708, T = 10090.21, t_MRCA = 2024-06-09, n0 = 32828.8, g = 0.0317, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.82, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5464, 5853, 9614], num_muts_ab counts: [9, 58, 13, 18, 4, 376, 47, 12, 60, 12, 90, 9]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.5 M steps/s, Step 2400000, log_posterior = -44981.97, log_G = -50681.44, log_coal = -973.60, log_other_priors = 6673.07, num_muts = 708, T = 9391.74, t_MRCA = 2024-06-26, n0 = 23554.7, g = 0.0321, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.41, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5856, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 376, 47, 14, 61, 12, 90, 7]\n",
            "0.6 M steps/s, Step 2500000, log_posterior = -44998.25, log_G = -50696.63, log_coal = -974.21, log_other_priors = 6672.60, num_muts = 708, T = 9588.25, t_MRCA = 2024-06-01, n0 = 35011.3, g = 0.0349, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.63, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5854, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 376, 47, 14, 59, 12, 90, 9]\n",
            "0.5 M steps/s, Step 2600000, log_posterior = -44971.53, log_G = -50682.81, log_coal = -961.30, log_other_priors = 6672.59, num_muts = 710, T = 9111.83, t_MRCA = 2024-07-14, n0 = 32150.8, g = 0.0371, mu = 1 * 10^-3 subst / site / year, M_ts = 573, M_tv = 137, kappa = 10, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 19, 2, 377, 47, 14, 59, 12, 91, 9]\n",
            "0.5 M steps/s, Step 2700000, log_posterior = -44965.95, log_G = -50676.44, log_coal = -961.03, log_other_priors = 6671.52, num_muts = 708, T = 9327.17, t_MRCA = 2024-07-26, n0 = 93793.5, g = 0.0478, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.32, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 375, 47, 14, 60, 12, 91, 8]\n",
            "0.5 M steps/s, Step 2800000, log_posterior = -44953.60, log_G = -50667.43, log_coal = -958.81, log_other_priors = 6672.64, num_muts = 710, T = 8891.98, t_MRCA = 2024-07-23, n0 = 50016.2, g = 0.0420, mu = 1 * 10^-3 subst / site / year, M_ts = 573, M_tv = 137, kappa = 7.29, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5854, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 378, 47, 14, 59, 12, 90, 9]\n",
            "0.6 M steps/s, Step 2900000, log_posterior = -44930.21, log_G = -50650.69, log_coal = -951.18, log_other_priors = 6671.66, num_muts = 708, T = 9026.10, t_MRCA = 2024-07-29, n0 = 84189.3, g = 0.0506, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.96, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5856, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 375, 47, 14, 61, 12, 91, 7]\n",
            "0.6 M steps/s, Step 3000000, log_posterior = -44940.32, log_G = -50656.24, log_coal = -956.39, log_other_priors = 6672.31, num_muts = 708, T = 9092.01, t_MRCA = 2024-07-19, n0 = 53369.2, g = 0.0444, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.37, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 375, 47, 14, 60, 12, 91, 8]\n",
            "0.6 M steps/s, Step 3100000, log_posterior = -44969.10, log_G = -50675.04, log_coal = -966.23, log_other_priors = 6672.16, num_muts = 708, T = 9309.16, t_MRCA = 2024-07-11, n0 = 46614.4, g = 0.0397, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 10.2, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8951, 5462, 5856, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 48, 14, 60, 12, 91, 8]\n",
            "0.5 M steps/s, Step 3200000, log_posterior = -44947.70, log_G = -50663.80, log_coal = -955.98, log_other_priors = 6672.08, num_muts = 708, T = 9172.40, t_MRCA = 2024-08-04, n0 = 50230.3, g = 0.0441, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.87, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5856, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 61, 12, 92, 7]\n",
            "0.5 M steps/s, Step 3300000, log_posterior = -44983.71, log_G = -50691.97, log_coal = -964.46, log_other_priors = 6672.72, num_muts = 708, T = 9505.63, t_MRCA = 2024-06-17, n0 = 37382.1, g = 0.0377, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.49, pi = [0.3, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5460, 5855, 9615], num_muts_ab counts: [10, 58, 13, 18, 2, 373, 47, 14, 60, 12, 93, 8]\n",
            "0.6 M steps/s, Step 3400000, log_posterior = -44935.76, log_G = -50649.68, log_coal = -958.00, log_other_priors = 6671.91, num_muts = 708, T = 8993.86, t_MRCA = 2024-07-29, n0 = 69699.7, g = 0.0472, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.87, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8954, 5461, 5854, 9614], num_muts_ab counts: [10, 59, 13, 18, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 3500000, log_posterior = -44930.52, log_G = -50645.91, log_coal = -957.34, log_other_priors = 6672.73, num_muts = 707, T = 9121.68, t_MRCA = 2024-07-20, n0 = 32274.2, g = 0.0384, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.13, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.6 M steps/s, Step 3600000, log_posterior = -44919.94, log_G = -50630.92, log_coal = -961.03, log_other_priors = 6672.01, num_muts = 707, T = 8887.74, t_MRCA = 2024-07-07, n0 = 52994.2, g = 0.0445, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.94, pi = [0.3, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [8, 58, 13, 19, 2, 373, 47, 15, 60, 12, 92, 8]\n",
            "0.6 M steps/s, Step 3700000, log_posterior = -44951.39, log_G = -50659.25, log_coal = -964.49, log_other_priors = 6672.34, num_muts = 707, T = 9269.57, t_MRCA = 2024-07-09, n0 = 39563.7, g = 0.0393, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 10.1, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.6 M steps/s, Step 3800000, log_posterior = -44908.08, log_G = -50624.87, log_coal = -955.92, log_other_priors = 6672.71, num_muts = 708, T = 8669.75, t_MRCA = 2024-07-24, n0 = 33732.5, g = 0.0391, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 8.96, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 61, 12, 93, 8]\n",
            "0.5 M steps/s, Step 3900000, log_posterior = -44887.35, log_G = -50608.11, log_coal = -951.35, log_other_priors = 6672.11, num_muts = 707, T = 8653.26, t_MRCA = 2024-07-30, n0 = 52351.1, g = 0.0445, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.49, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5463, 5855, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 12, 91, 8]\n",
            "0.5 M steps/s, Step 4000000, log_posterior = -44894.48, log_G = -50618.23, log_coal = -948.18, log_other_priors = 6671.93, num_muts = 707, T = 8764.21, t_MRCA = 2024-08-03, n0 = 66442.2, g = 0.0479, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.97, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.6 M steps/s, Step 4100000, log_posterior = -44869.75, log_G = -50592.15, log_coal = -950.30, log_other_priors = 6672.70, num_muts = 707, T = 8456.32, t_MRCA = 2024-08-09, n0 = 34920.8, g = 0.0419, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.69, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5854, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 59, 12, 92, 9]\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 0.494212 days\n",
            "\n",
            "*******\n",
            "\n",
            "0.5 M steps/s, Step 4200000, log_posterior = -44890.82, log_G = -50616.40, log_coal = -946.84, log_other_priors = 6672.43, num_muts = 709, T = 8415.15, t_MRCA = 2024-08-03, n0 = 51136.3, g = 0.0473, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 7.84, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5855, 9616], num_muts_ab counts: [9, 59, 13, 19, 2, 371, 48, 14, 60, 12, 94, 8]\n",
            "0.6 M steps/s, Step 4300000, log_posterior = -44889.35, log_G = -50610.27, log_coal = -951.54, log_other_priors = 6672.46, num_muts = 707, T = 8679.71, t_MRCA = 2024-08-02, n0 = 40556.2, g = 0.0422, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.11, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [9, 58, 13, 19, 2, 371, 47, 14, 59, 12, 94, 9]\n",
            "0.6 M steps/s, Step 4400000, log_posterior = -44904.63, log_G = -50625.40, log_coal = -951.85, log_other_priors = 6672.61, num_muts = 708, T = 8667.69, t_MRCA = 2024-07-11, n0 = 33013.5, g = 0.0404, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.52, pi = [0.3, 0.19, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5853, 9617], num_muts_ab counts: [9, 58, 13, 19, 3, 371, 47, 13, 59, 12, 95, 9]\n",
            "0.5 M steps/s, Step 4500000, log_posterior = -44920.44, log_G = -50643.87, log_coal = -948.94, log_other_priors = 6672.37, num_muts = 708, T = 8910.57, t_MRCA = 2024-08-08, n0 = 53944.7, g = 0.0452, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 7.97, pi = [0.3, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5854, 9617], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 59, 12, 94, 9]\n",
            "0.6 M steps/s, Step 4600000, log_posterior = -44911.18, log_G = -50631.93, log_coal = -951.58, log_other_priors = 6672.33, num_muts = 707, T = 8936.11, t_MRCA = 2024-07-25, n0 = 56474.0, g = 0.0462, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 7.9, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 4700000, log_posterior = -44924.33, log_G = -50641.57, log_coal = -954.92, log_other_priors = 6672.16, num_muts = 707, T = 9024.51, t_MRCA = 2024-07-06, n0 = 49867.6, g = 0.0431, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.54, pi = [0.3, 0.18, 0.2, 0.32], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5460, 5855, 9616], num_muts_ab counts: [9, 58, 13, 19, 2, 371, 47, 14, 60, 12, 94, 8]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.5 M steps/s, Step 4800000, log_posterior = -44908.50, log_G = -50627.58, log_coal = -953.21, log_other_priors = 6672.29, num_muts = 708, T = 8715.01, t_MRCA = 2024-07-23, n0 = 52618.6, g = 0.0450, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.52, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5854, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 59, 12, 92, 9]\n",
            "0.5 M steps/s, Step 4900000, log_posterior = -44942.98, log_G = -50650.19, log_coal = -965.43, log_other_priors = 6672.63, num_muts = 707, T = 9140.21, t_MRCA = 2024-07-31, n0 = 44889.5, g = 0.0378, mu = 1 * 10^-3 subst / site / year, M_ts = 569, M_tv = 138, kappa = 8.05, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 19, 3, 373, 47, 14, 59, 12, 92, 9]\n",
            "0.5 M steps/s, Step 5000000, log_posterior = -44916.31, log_G = -50631.18, log_coal = -957.79, log_other_priors = 6672.66, num_muts = 708, T = 8762.34, t_MRCA = 2024-07-12, n0 = 47360.3, g = 0.0420, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 7.43, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5853, 9616], num_muts_ab counts: [9, 58, 13, 18, 3, 374, 47, 13, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 5100000, log_posterior = -44936.57, log_G = -50647.62, log_coal = -961.40, log_other_priors = 6672.46, num_muts = 707, T = 9130.27, t_MRCA = 2024-07-02, n0 = 40628.9, g = 0.0413, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.19, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 5200000, log_posterior = -44911.41, log_G = -50620.17, log_coal = -964.22, log_other_priors = 6672.97, num_muts = 707, T = 8763.05, t_MRCA = 2024-07-09, n0 = 22426.7, g = 0.0350, mu = 1 * 10^-3 subst / site / year, M_ts = 569, M_tv = 138, kappa = 10, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 18, 3, 373, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 5300000, log_posterior = -44925.25, log_G = -50628.84, log_coal = -969.85, log_other_priors = 6673.45, num_muts = 707, T = 8855.50, t_MRCA = 2024-06-24, n0 = 23219.1, g = 0.0332, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 7.57, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 60, 12, 92, 9]\n",
            "0.5 M steps/s, Step 5400000, log_posterior = -44944.86, log_G = -50651.75, log_coal = -965.74, log_other_priors = 6672.64, num_muts = 707, T = 9179.68, t_MRCA = 2024-06-26, n0 = 31898.6, g = 0.0352, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.88, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5463, 5855, 9613], num_muts_ab counts: [9, 58, 13, 18, 3, 374, 47, 13, 61, 12, 91, 8]\n",
            "0.5 M steps/s, Step 5500000, log_posterior = -44973.61, log_G = -50676.43, log_coal = -969.35, log_other_priors = 6672.17, num_muts = 707, T = 9483.95, t_MRCA = 2024-07-07, n0 = 50314.0, g = 0.0378, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.79, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 374, 47, 14, 60, 12, 91, 9]\n",
            "0.5 M steps/s, Step 5600000, log_posterior = -44932.34, log_G = -50647.57, log_coal = -956.51, log_other_priors = 6671.74, num_muts = 707, T = 9116.35, t_MRCA = 2024-07-10, n0 = 69261.4, g = 0.0458, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.89, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5463, 5854, 9614], num_muts_ab counts: [9, 58, 13, 18, 4, 373, 47, 12, 61, 12, 92, 8]\n",
            "0.5 M steps/s, Step 5700000, log_posterior = -44941.19, log_G = -50655.52, log_coal = -957.83, log_other_priors = 6672.17, num_muts = 707, T = 9233.19, t_MRCA = 2024-07-29, n0 = 49674.8, g = 0.0419, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.63, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 374, 47, 14, 60, 12, 91, 9]\n",
            "0.5 M steps/s, Step 5800000, log_posterior = -44941.00, log_G = -50652.70, log_coal = -960.43, log_other_priors = 6672.12, num_muts = 707, T = 9186.02, t_MRCA = 2024-07-06, n0 = 63731.5, g = 0.0431, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.49, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 374, 47, 14, 60, 12, 91, 9]\n",
            "0.5 M steps/s, Step 5900000, log_posterior = -44920.07, log_G = -50629.87, log_coal = -963.67, log_other_priors = 6673.46, num_muts = 707, T = 8898.78, t_MRCA = 2024-07-10, n0 = 20374.9, g = 0.0327, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.13, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 6000000, log_posterior = -44946.89, log_G = -50663.99, log_coal = -954.66, log_other_priors = 6671.76, num_muts = 708, T = 9173.58, t_MRCA = 2024-07-28, n0 = 75033.7, g = 0.0469, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.28, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 18, 2, 374, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 6100000, log_posterior = -44903.16, log_G = -50619.76, log_coal = -955.64, log_other_priors = 6672.25, num_muts = 707, T = 8800.74, t_MRCA = 2024-07-16, n0 = 48783.8, g = 0.0438, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.19, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5853, 9617], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 58, 12, 92, 11]\n",
            "0.5 M steps/s, Step 6200000, log_posterior = -44896.38, log_G = -50609.49, log_coal = -960.07, log_other_priors = 6673.19, num_muts = 707, T = 8667.15, t_MRCA = 2024-07-04, n0 = 21904.3, g = 0.0355, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.97, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5853, 9617], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 58, 12, 92, 11]\n",
            "0.6 M steps/s, Step 6300000, log_posterior = -44914.05, log_G = -50631.45, log_coal = -955.25, log_other_priors = 6672.65, num_muts = 708, T = 8735.60, t_MRCA = 2024-07-23, n0 = 42486.0, g = 0.0433, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 7.9, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 18, 3, 374, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 6400000, log_posterior = -44914.23, log_G = -50637.35, log_coal = -948.85, log_other_priors = 6671.97, num_muts = 707, T = 9006.10, t_MRCA = 2024-07-28, n0 = 61958.9, g = 0.0486, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.09, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5462, 5852, 9616], num_muts_ab counts: [9, 59, 13, 18, 3, 373, 46, 13, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 6500000, log_posterior = -44889.99, log_G = -50612.01, log_coal = -949.98, log_other_priors = 6672.00, num_muts = 707, T = 8659.16, t_MRCA = 2024-08-05, n0 = 59192.6, g = 0.0490, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.15, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 18, 2, 373, 47, 14, 59, 12, 92, 10]\n",
            "0.5 M steps/s, Step 6600000, log_posterior = -44876.38, log_G = -50596.72, log_coal = -952.34, log_other_priors = 6672.69, num_muts = 708, T = 8356.47, t_MRCA = 2024-08-05, n0 = 34096.0, g = 0.0412, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.92, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 18, 2, 375, 47, 14, 60, 12, 91, 9]\n",
            "0.5 M steps/s, Step 6700000, log_posterior = -44903.93, log_G = -50619.86, log_coal = -956.65, log_other_priors = 6672.58, num_muts = 708, T = 8635.99, t_MRCA = 2024-07-17, n0 = 33278.2, g = 0.0410, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.59, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 375, 47, 14, 60, 12, 91, 8]\n",
            "0.5 M steps/s, Step 6800000, log_posterior = -44937.55, log_G = -50649.53, log_coal = -960.64, log_other_priors = 6672.62, num_muts = 709, T = 8823.39, t_MRCA = 2024-07-13, n0 = 26028.2, g = 0.0380, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 10.9, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5461, 5855, 9614], num_muts_ab counts: [10, 58, 13, 17, 2, 376, 47, 14, 60, 13, 91, 8]\n",
            "0.5 M steps/s, Step 6900000, log_posterior = -44911.92, log_G = -50629.25, log_coal = -955.13, log_other_priors = 6672.46, num_muts = 707, T = 8912.83, t_MRCA = 2024-07-20, n0 = 43240.9, g = 0.0424, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.79, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 12, 91, 8]\n",
            "0.5 M steps/s, Step 7000000, log_posterior = -44894.29, log_G = -50607.86, log_coal = -959.15, log_other_priors = 6672.72, num_muts = 707, T = 8656.42, t_MRCA = 2024-07-10, n0 = 34683.2, g = 0.0418, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.63, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.4 M steps/s, Step 7100000, log_posterior = -44905.13, log_G = -50623.77, log_coal = -953.54, log_other_priors = 6672.18, num_muts = 707, T = 8838.75, t_MRCA = 2024-08-08, n0 = 62564.4, g = 0.0454, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.15, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 12, 91, 8]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.3 M steps/s, Step 7200000, log_posterior = -44894.89, log_G = -50615.95, log_coal = -950.92, log_other_priors = 6671.99, num_muts = 707, T = 8734.14, t_MRCA = 2024-08-03, n0 = 51660.6, g = 0.0442, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 10.2, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5856, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 61, 12, 91, 7]\n",
            "0.3 M steps/s, Step 7300000, log_posterior = -44914.81, log_G = -50630.90, log_coal = -956.04, log_other_priors = 6672.13, num_muts = 707, T = 8937.12, t_MRCA = 2024-07-17, n0 = 63669.5, g = 0.0456, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.33, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5463, 5855, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 375, 47, 14, 60, 12, 90, 8]\n",
            "0.3 M steps/s, Step 7400000, log_posterior = -44897.99, log_G = -50622.49, log_coal = -948.06, log_other_priors = 6672.56, num_muts = 707, T = 8815.58, t_MRCA = 2024-07-23, n0 = 45156.4, g = 0.0450, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 7.96, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5461, 5855, 9614], num_muts_ab counts: [10, 58, 13, 18, 2, 374, 47, 14, 60, 12, 91, 8]\n",
            "0.3 M steps/s, Step 7500000, log_posterior = -44912.89, log_G = -50631.33, log_coal = -954.02, log_other_priors = 6672.46, num_muts = 708, T = 8788.03, t_MRCA = 2024-07-20, n0 = 39239.6, g = 0.0403, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.44, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 13, 92, 7]\n",
            "0.3 M steps/s, Step 7600000, log_posterior = -44884.57, log_G = -50608.83, log_coal = -948.76, log_other_priors = 6673.02, num_muts = 708, T = 8431.93, t_MRCA = 2024-07-25, n0 = 36443.8, g = 0.0437, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 138, kappa = 6.92, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 57, 13, 19, 3, 374, 47, 14, 60, 13, 92, 7]\n",
            "0.3 M steps/s, Step 7700000, log_posterior = -44919.35, log_G = -50632.61, log_coal = -959.73, log_other_priors = 6672.99, num_muts = 708, T = 8802.78, t_MRCA = 2024-07-14, n0 = 27629.3, g = 0.0360, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.78, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 13, 92, 7]\n",
            "0.3 M steps/s, Step 7800000, log_posterior = -44914.69, log_G = -50637.94, log_coal = -948.69, log_other_priors = 6671.94, num_muts = 708, T = 8864.13, t_MRCA = 2024-08-07, n0 = 65727.0, g = 0.0477, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 8.98, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 13, 93, 7]\n",
            "0.3 M steps/s, Step 7900000, log_posterior = -44896.96, log_G = -50616.52, log_coal = -952.90, log_other_priors = 6672.45, num_muts = 709, T = 8451.32, t_MRCA = 2024-08-05, n0 = 43090.9, g = 0.0418, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 8.88, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 59, 13, 93, 8]\n",
            "0.3 M steps/s, Step 8000000, log_posterior = -44906.25, log_G = -50637.41, log_coal = -940.89, log_other_priors = 6672.05, num_muts = 709, T = 8698.02, t_MRCA = 2024-07-31, n0 = 69969.2, g = 0.0494, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 8.04, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5856, 9613], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 48, 14, 61, 13, 92, 6]\n",
            "0.4 M steps/s, Step 8100000, log_posterior = -44867.79, log_G = -50599.79, log_coal = -939.81, log_other_priors = 6671.81, num_muts = 709, T = 8238.56, t_MRCA = 2024-08-19, n0 = 91766.8, g = 0.0564, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 7.5, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 19, 2, 375, 47, 14, 59, 13, 93, 8]\n",
            "0.5 M steps/s, Step 8200000, log_posterior = -44877.26, log_G = -50599.80, log_coal = -949.60, log_other_priors = 6672.13, num_muts = 708, T = 8399.37, t_MRCA = 2024-08-03, n0 = 52013.8, g = 0.0462, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.31, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 19, 2, 373, 47, 14, 59, 13, 94, 8]\n",
            "0.5 M steps/s, Step 8300000, log_posterior = -44898.40, log_G = -50626.09, log_coal = -943.74, log_other_priors = 6671.43, num_muts = 708, T = 8673.48, t_MRCA = 2024-08-04, n0 = 63916.3, g = 0.0504, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 11.8, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 57, 13, 19, 2, 373, 47, 14, 59, 13, 94, 8]\n",
            "0.5 M steps/s, Step 8400000, log_posterior = -44890.54, log_G = -50621.48, log_coal = -941.14, log_other_priors = 6672.08, num_muts = 708, T = 8631.00, t_MRCA = 2024-08-04, n0 = 67794.0, g = 0.0518, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 7.92, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 59, 13, 94, 8]\n",
            "0.5 M steps/s, Step 8500000, log_posterior = -44894.27, log_G = -50619.19, log_coal = -946.47, log_other_priors = 6671.39, num_muts = 707, T = 8787.19, t_MRCA = 2024-08-09, n0 = 111601.2, g = 0.0566, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.54, pi = [0.29, 0.18, 0.2, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 8600000, log_posterior = -44888.99, log_G = -50610.24, log_coal = -950.83, log_other_priors = 6672.08, num_muts = 707, T = 8664.24, t_MRCA = 2024-07-29, n0 = 71979.6, g = 0.0502, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 7.74, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 8700000, log_posterior = -44881.54, log_G = -50601.36, log_coal = -952.39, log_other_priors = 6672.21, num_muts = 708, T = 8402.02, t_MRCA = 2024-08-02, n0 = 50445.9, g = 0.0457, mu = 1 * 10^-3 subst / site / year, M_ts = 571, M_tv = 137, kappa = 9.1, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 8800000, log_posterior = -44865.75, log_G = -50592.83, log_coal = -945.23, log_other_priors = 6672.31, num_muts = 707, T = 8367.52, t_MRCA = 2024-08-01, n0 = 44796.7, g = 0.0454, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.2, pi = [0.3, 0.19, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 8900000, log_posterior = -44898.53, log_G = -50614.05, log_coal = -956.79, log_other_priors = 6672.31, num_muts = 707, T = 8708.48, t_MRCA = 2024-07-16, n0 = 41759.8, g = 0.0426, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.76, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 9000000, log_posterior = -44910.62, log_G = -50623.57, log_coal = -959.71, log_other_priors = 6672.67, num_muts = 707, T = 8826.37, t_MRCA = 2024-07-18, n0 = 39817.0, g = 0.0388, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.38, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 9100000, log_posterior = -44885.67, log_G = -50605.26, log_coal = -953.12, log_other_priors = 6672.71, num_muts = 707, T = 8618.66, t_MRCA = 2024-07-30, n0 = 29911.3, g = 0.0412, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.45, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 9200000, log_posterior = -44905.09, log_G = -50624.23, log_coal = -953.28, log_other_priors = 6672.42, num_muts = 707, T = 8806.21, t_MRCA = 2024-07-16, n0 = 39968.2, g = 0.0421, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.41, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5855, 9615], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 60, 12, 93, 8]\n",
            "0.5 M steps/s, Step 9300000, log_posterior = -44888.73, log_G = -50615.91, log_coal = -945.20, log_other_priors = 6672.38, num_muts = 707, T = 8755.67, t_MRCA = 2024-08-07, n0 = 46668.8, g = 0.0438, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.74, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5856, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 372, 47, 14, 61, 12, 93, 7]\n",
            "0.5 M steps/s, Step 9400000, log_posterior = -44881.10, log_G = -50604.23, log_coal = -949.28, log_other_priors = 6672.41, num_muts = 707, T = 8605.81, t_MRCA = 2024-08-04, n0 = 43623.4, g = 0.0433, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.94, pi = [0.29, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.5 M steps/s, Step 9500000, log_posterior = -44864.65, log_G = -50598.58, log_coal = -937.91, log_other_priors = 6671.84, num_muts = 707, T = 8459.02, t_MRCA = 2024-08-03, n0 = 79143.5, g = 0.0545, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.15, pi = [0.29, 0.18, 0.19, 0.34], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 9600000, log_posterior = -44841.70, log_G = -50569.22, log_coal = -944.35, log_other_priors = 6671.87, num_muts = 707, T = 8157.35, t_MRCA = 2024-08-06, n0 = 57627.5, g = 0.0495, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.95, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 60, 12, 92, 8]\n",
            "0.5 M steps/s, Step 9700000, log_posterior = -44915.10, log_G = -50642.45, log_coal = -944.55, log_other_priors = 6671.90, num_muts = 709, T = 8737.24, t_MRCA = 2024-07-30, n0 = 70466.8, g = 0.0492, mu = 1 * 10^-3 subst / site / year, M_ts = 572, M_tv = 137, kappa = 8.77, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8953, 5460, 5854, 9616], num_muts_ab counts: [10, 58, 13, 18, 2, 373, 47, 14, 59, 12, 94, 9]\n",
            "0.5 M steps/s, Step 9800000, log_posterior = -44868.81, log_G = -50601.55, log_coal = -938.81, log_other_priors = 6671.55, num_muts = 707, T = 8559.92, t_MRCA = 2024-08-09, n0 = 83723.9, g = 0.0538, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 9.38, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5854, 9616], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 59, 12, 92, 9]\n",
            "0.5 M steps/s, Step 9900000, log_posterior = -44880.24, log_G = -50615.57, log_coal = -935.60, log_other_priors = 6670.93, num_muts = 707, T = 8745.43, t_MRCA = 2024-08-20, n0 = 173815.5, g = 0.0619, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.33, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5461, 5853, 9617], num_muts_ab counts: [9, 58, 13, 19, 2, 373, 47, 14, 58, 12, 92, 10]\n",
            "0.5 M steps/s, Step 10000000, log_posterior = -44862.75, log_G = -50594.87, log_coal = -939.57, log_other_priors = 6671.70, num_muts = 707, T = 8453.12, t_MRCA = 2024-08-07, n0 = 85128.6, g = 0.0545, mu = 1 * 10^-3 subst / site / year, M_ts = 570, M_tv = 137, kappa = 8.49, pi = [0.3, 0.18, 0.19, 0.33], alpha OFF, nu = 1.000, Root Seq counts: [8952, 5462, 5855, 9614], num_muts_ab counts: [9, 58, 13, 19, 2, 374, 47, 14, 60, 12, 91, 8]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "<h1>All done! 🎉</h1>\n",
              "<h2>To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.</h2>\n",
              "<h2>To further visualize your Delphy output, upload the <code>.dphy</code> (and <code>metadata.csv</code>) file(s) to <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a></h2>\n"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 10.5 s, sys: 1.7 s, total: 12.2 s\n",
            "Wall time: 4min 25s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ihN2Y1S1cSYC"
      },
      "execution_count": 4,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}